qa: address PMD GuardLogStatement warnings

soloturn opened the pull request:
using fluent logger avoids evaluating methods, and fixes the PMD warning: log not surrounded by "if". this is in case the argument to a parameterized logger is a method call, like here:
```diff
- logger.error("onAudioEnd() notification failed for {}", entry.getValue(), e);
+ logger.atError().addArgument(() -> entry.getValue()).addArgument(e).log("onAudioEnd() notification failed for {}");
```

Edit by niruandaleth/jdrueckert:
PMD warnings were fixed in different ways, among others the following variant of the fluent API which is less verbose than the one originally proposed (see above):
```diff
- logger.error("onAudioEnd() notification failed for {}", entry.getValue(), e);
+ logger.atError().log("onAudioEnd() notification failed for {}", entry.getValue(), e);
```

soloturn committed:
qa engine, audio, fluent logger.

soloturn committed:
qa engine, config, fluent logger.

soloturn committed:
qa engine, core, fluent logger.

soloturn committed:
qa engine, entitySystem, fluent logger.

soloturn committed:
qa engine, game, fluent logger.

soloturn committed:
qa engine, identity, fluent logger.

soloturn committed:
qa engine, input, fluent logger.

soloturn committed:
qa engine, logic, fluent logger.

soloturn committed:
qa engine, network, fluent logger.

soloturn committed:
qa engine, particles, fluent logger.

soloturn committed:
qa engine, persistence, fluent logger.

soloturn committed:
qa engine, physics, fluent logger.

soloturn committed:
qa engine, recording, fluent logger.

soloturn committed:
qa engine, rendering, fluent logger.

soloturn committed:
qa engine, registry, fluent logger.

soloturn committed:
qa engine, utilities, fluent logger.

soloturn committed:
qa engine, world, fluent logger.

BenjaminAmos commented:
This might take a while to review (92 files changed) but at a cursory glance there's a few questions that occurred to me about this:
 - For the PMD warning (log not surrounded by "if"), is this about still evaluating methods even when logging is disabled? Is there a significant impact already from not doing this?
 - The "fluent" logger style appears more verbose and makes logging harder to distinguish from existing code. Is this to be used everywhere or just where a method is called for a log parameter?
 - How would you suggest we enforce this change? I suspect that new code will bring new log statements and they are highly likely to use the original logger methods unless enforced.

soloturn commented:
> * For the PMD warning (log not surrounded by "if"), is this about still evaluating methods even when logging is disabled? Is there a significant impact already from not doing this?

true. methods still would be called. depending which method it is the impact varies. not calling the method when not logging has, contrary, a clear and predictable impact: no code executed. which is good.

> * The "fluent" logger style appears more verbose and makes logging harder to distinguish from existing code. Is this to be used everywhere or just where a method is called for a log parameter?

i changed it only where a method is called as log parameter and pmd warned.

> * How would you suggest we enforce this change? I suspect that new code will bring new log statements and they are highly likely to use the original logger methods unless enforced.

we could turn on pmd again, it warns or blocks depending how we set it. there is no problem to use original logger methods, at least imo. there are other opinionated approaches, like using [google flogger](https://github.com/google/flogger), which would convert everything, but that might be too much for terasology. personally i would not enforce it, but rely on peoples working brains on code review. as well as local ```gradle clean build``` warning about it. but that is not my call, i am happy with any approach you guys take.

soloturn commented:
> This might take a while to review (92 files changed) 

you find it easier to chunk it?

jdrueckert commented on engine/src/main/java/org/terasology/engine/audio/openAL/OpenALManager.java:
```diff
@@ -64,15 +64,16 @@ public OpenALManager(AudioConfig config) throws OpenALException {
         AL.createCapabilities(alcCapabilities);
 
-        logger.info("OpenAL {} initialized!", AL10.alGetString(AL10.AL_VERSION));
+        logger.atInfo().addArgument(() -> AL10.alGetString(AL10.AL_VERSION)).log("OpenAL {} initialized!");
```
The likelihood of these info logs not being written is comparably low as they are on info level and relevant for audio-related troubleshooting. Also, they are only written once when constructing the OpenALManager, so performance impact is limited.
Please ignore PMD.GuardLogStatement issues for this constructor by adding the following annotation above it:
```java
@SuppressWarnings("PMD.GuardLogStatement")
```

jdrueckert replied:
looks like you missed one there when reverting

jdrueckert commented on engine/src/main/java/org/terasology/engine/audio/openAL/OpenALManager.java:
```diff
@@ -260,7 +261,7 @@ private void notifyEndListeners(boolean interrupted) {
                     entry.getValue().onAudioEnd(interrupted);
                 } catch (Exception e) {
-                    logger.error("onAudioEnd() notification failed for {}", entry.getValue(), e);
+                    logger.atError().addArgument(() -> entry.getValue()).addArgument(e).log("onAudioEnd() notification failed for {}");
```
The associated method will only be evaluated if the catch clause is evaluated, so I think we do not need the fluent API here. Please ignore PMD warnings for this line instead:
```suggestion
                    logger.error("onAudioEnd() notification failed for {}", entry.getValue(), e); //NOPMD
```

soloturn replied:
fixed

jdrueckert replied:
this doesn't seem to be fixed :thinking:

jdrueckert commented on engine/src/main/java/org/terasology/engine/audio/openAL/staticSound/OpenALSound.java:
```diff
@@ -104,7 +104,7 @@ protected void doReload(StaticSoundData newData) {
             });
         } catch (InterruptedException e) {
-            logger.error("Failed to reload {}", getUrn(), e);
+            logger.atError().addArgument(() -> getUrn()).addArgument(e).log("Failed to reload {}");
```
see above
```suggestion
            logger.error("Failed to reload {}", getUrn(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/audio/openAL/streamingSound/OpenALStreamingSound.java:
```diff
@@ -123,7 +123,7 @@ protected void doReload(StreamingSoundData data) {
             GameThread.synch(this::initializeBuffers);
         } catch (InterruptedException e) {
-            logger.error("Failed to reload {}", getUrn(), e);
+            logger.atError().addArgument(() -> getUrn()).addArgument(e).log("Failed to reload {}");
```
see above
```suggestion
            logger.error("Failed to reload {}", getUrn(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/config/flexible/AutoConfigManager.java:
```diff
@@ -85,7 +85,7 @@ private <T extends AutoConfig> void loadSettingsFromDisk(Class<T> configClass, T
             mergeConfig(configClass, loadedConfig, config);
         } catch (Exception e) {
-            logger.error("Error while loading config {} from disk", config.getId(), e);
+            logger.atError().addArgument(() -> config.getId()).addArgument(e).log("Error while loading config {} from disk");
```
see above
```suggestion
            logger.error("Error while loading config {} from disk", config.getId(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/config/flexible/AutoConfigManager.java:
```diff
@@ -116,7 +116,7 @@ private void saveConfigToDisk(AutoConfig config) {
             serializer.serialize(config, TypeInfo.of((Class<AutoConfig>) config.getClass()), output);
         } catch (IOException e) {
-            logger.error("Error while saving config {} to disk", config.getId(), e);
+            logger.atError().addArgument(() -> config.getId()).addArgument(e).log("Error while saving config {} to disk");
```
see above
```suggestion
            logger.error("Error while saving config {} to disk", config.getId(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/TerasologyEngine.java:
```diff
@@ -543,7 +547,7 @@ public void cleanup() {
                 subsystem.preShutdown();
             } catch (RuntimeException e) {
-                logger.error("Error preparing to shutdown {} subsystem", subsystem.getName(), e);
+                logger.atError().addArgument(() -> subsystem.getName()).addArgument(e).log("Error preparing to shutdown {} subsystem");
```
```suggestion
                logger.error("Error preparing to shutdown {} subsystem", subsystem.getName(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/TerasologyEngine.java:
```diff
@@ -553,7 +557,7 @@ public void cleanup() {
                 subsystem.shutdown();
             } catch (RuntimeException e) {
-                logger.error("Error shutting down {} subsystem", subsystem.getName(), e);
+                logger.atError().addArgument(() -> subsystem.getName()).addArgument(e).log("Error shutting down {} subsystem");
```
```suggestion
                logger.error("Error shutting down {} subsystem", subsystem.getName(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/TerasologyEngine.java:
```diff
@@ -266,16 +266,20 @@ private void verifyInitialisation() {
+                addArgument(() -> System.getProperty("os.version")).
+                log("OS: {}, arch: {}, version: {}");
+        logger.atInfo().addArgument(() -> Runtime.getRuntime().maxMemory() / ONE_MEBIBYTE).log("Max. Memory: {} MiB");
+        logger.atInfo().addArgument(() -> Runtime.getRuntime().availableProcessors()).log("Processors: {}");
```
Ignore PMD GuardLogStatement warnings in `logEnvironmentInfo` instead using the following annotation:
```java
@SuppressWarnings("PMD.GuardLogStatement")
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/TerasologyEngine.java:
```diff
@@ -253,7 +253,7 @@ public void initialize() {
 
         double seconds = 0.001 * totalInitTime.elapsed(TimeUnit.MILLISECONDS);
-        logger.info("Initialization completed in {}sec.", String.format("%.2f", seconds));
+        logger.atInfo().addArgument(() -> String.format("%.2f", seconds)).log("Initialization completed in {}sec.");
```
`String.format` complexity and impact are neglectable.
```suggestion
        logger.info("Initialization completed in {}sec.", String.format("%.2f", seconds)); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseWorld.java:
```diff
@@ -104,8 +104,8 @@ public boolean step() {
-            logger.error("Unable to load world generator {}. Available world generators: {}",
-                    worldInfo.getWorldGenerator(), worldGeneratorManager.getWorldGenerators());
+            logger.atError().addArgument(() -> worldInfo.getWorldGenerator()).addArgument(() -> worldGeneratorManager.getWorldGenerators()).
+                    log("Unable to load world generator {}. Available world generators: {}");
```
```suggestion
            logger.error("Unable to load world generator {}. Available world generators: {}",
                    worldInfo.getWorldGenerator(), worldGeneratorManager.getWorldGenerators()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/module/ModuleInstaller.java:
```diff
@@ -48,7 +48,7 @@ public List<Module> call() throws Exception {
                 newInstalledModules.add(module);
             } catch (IOException e) {
-                logger.warn("Could not load module {}", filePath.getFileName(), e);
+                logger.atWarn().addArgument(() -> filePath.getFileName()).addArgument(e).log("Could not load module {}");
```
```suggestion
                logger.warn("Could not load module {}", filePath.getFileName(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/subsystem/lwjgl/GLFWErrorCallback.java:
```diff
@@ -15,6 +15,7 @@ public class GLFWErrorCallback implements GLFWErrorCallbackI {
     @Override
     public void invoke(int error, long description) {
-        logger.error("Received error. Code: {}, Description: {}", error, MemoryUtil.memASCII(description));
+        logger.atError().addArgument(() -> error).addArgument(() -> MemoryUtil.memASCII(description)).
```
Only evaluated if this method is evaluated in which case it is relevant and should be.
Ignore via annotation:
```java
@SuppressWarnings("PMD.GuardLogStatement")
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseWorld.java:
```diff
@@ -93,7 +93,7 @@ public boolean step() {
         }
 
-        logger.info("World seed: \"{}\"", worldInfo.getSeed());
+        logger.atInfo().addArgument(() -> worldInfo.getSeed()).log("World seed: \"{}\"");
```
Complexity and impact of `getSeed()` are neglectable, ignore using comment:
```suggestion
        logger.info("World seed: \"{}\"", worldInfo.getSeed()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/config/flexible/constraints/StringConstraint.java:
```diff
@@ -50,10 +50,12 @@ public boolean isSatisfiedBy(String value) {
                         .map(StringConstraint::getDescription)
-                        .collect(Collectors.joining(",", "[", "]")));
+                        .collect(Collectors.joining(",", "[", "]"))).
+                log("String [{}] does not match the conditions: {}");
```
Only evaluated if this method is evaluated. Purpose of the method is to log.
Ignore using annotation:
```java
@SuppressWarnings("PMD.GuardLogStatement")
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/config/flexible/constraints/NumberRangeConstraint.java:
```diff
@@ -97,7 +97,12 @@ public boolean isSatisfiedBy(T value) {
+                addArgument(min != null ? min : "UNBOUNDED").
+                addArgument(max != null ? max : "UNBOUNDED").
+                addArgument(maxInclusive ? "]" : ")").
+                log("Value {} is not in the range {}{}, {}{}");
```
Only evaluated if this method is evaluated. Purpose of the method is to log.
Ignore using annotation:
```java
@SuppressWarnings("PMD.GuardLogStatement")
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/config/flexible/constraints/LocaleConstraint.java:
```diff
@@ -32,11 +32,11 @@ public boolean isSatisfiedBy(Locale value) {
-                        .collect(Collectors.joining(",", "[", "]"))
-        );
+                        .collect(Collectors.joining(",", "[", "]"))).
+                log("Locale {} should be one of {}");
```
Only evaluated if this method is evaluated. Purpose of the method is to log.
Ignore using annotation:
```java
@SuppressWarnings("PMD.GuardLogStatement")
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/entitySystem/metadata/ComponentLibrary.java:
```diff
@@ -46,11 +46,13 @@ public <T> ComponentLibrary createCopyUsingCopyStrategy(Class<T> type, CopyStrat
         } catch (NoSuchMethodException e) {
-            logger.error("Unable to register class {}: Default Constructor Required", type.getSimpleName(), e);
+            logger.atError().addArgument(() -> type.getSimpleName()).addArgument(() -> e).
+                    log("Unable to register class {}: Default Constructor Required");
```
```suggestion
            logger.error("Unable to register class {}: Default Constructor Required", type.getSimpleName(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/entitySystem/metadata/ComponentLibrary.java:
```diff
@@ -46,11 +46,13 @@ public <T> ComponentLibrary createCopyUsingCopyStrategy(Class<T> type, CopyStrat
             // log what class was not found so that diagnosis is easier
-            logger.error("Class not found, {}", type.getSimpleName(), e);
+            logger.atError().addArgument(() -> type.getSimpleName()).addArgument(() -> e).
+                    log("Class not found, {}");
```
```suggestion
            logger.error("Class not found, {}", type.getSimpleName(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/entitySystem/metadata/EventLibrary.java:
```diff
@@ -30,7 +30,8 @@ public EventLibrary(ModuleEnvironment environment, ReflectFactory reflectFactory
         } catch (NoSuchMethodException e) {
-            logger.error("Unable to register class {}: Default Constructor Required", type.getSimpleName(), e);
+            logger.atError().addArgument(() -> type.getSimpleName()).addArgument(() -> e).
+                    log("Unable to register class {}: Default Constructor Required");
```
```suggestion
            logger.error("Unable to register class {}: Default Constructor Required", type.getSimpleName(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/behavior/actions/conditions/ConditionAction.java:
```diff
@@ -98,7 +98,7 @@ protected boolean condition(Actor actor) throws ClassNotFoundException, NoSuchFi
                                 break;
                             default:
-                                logger.error("Unsupported guard value type: {}", tokens[0]);
+                                logger.atError().addArgument(() -> tokens[0]).log("Unsupported guard value type: {}");
```
The associated method is expected to only be evaluated if the default case applies.
In this case the log is required, so the evaluation is not undesired and the PMD warning should be ignored:
```suggestion
                                logger.error("Unsupported guard value type: {}", tokens[0]); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/behavior/actions/conditions/ConditionAction.java:
```diff
@@ -115,7 +115,7 @@ protected boolean condition(Actor actor) throws ClassNotFoundException, NoSuchFi
                                     break;
                                 default:
-                                    logger.error("Unsupported operation for boolean values: {}", tokens[2]);
+                                    logger.atError().addArgument(() -> tokens[2]).log("Unsupported operation for boolean values: {}");
```
same here
```suggestion
                                    logger.error("Unsupported operation for boolean values: {}", tokens[2]); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/behavior/actions/conditions/ConditionAction.java:
```diff
@@ -142,7 +142,7 @@ protected boolean condition(Actor actor) throws ClassNotFoundException, NoSuchFi
                                     break;
                                 default:
-                                    logger.error("Unsupported operation for numeric values: {}", tokens[2]);
+                                    logger.atError().addArgument(() -> tokens[2]).log("Unsupported operation for numeric values: {}");
```
same here
```suggestion
                                    logger.error("Unsupported operation for numeric values: {}", tokens[2]); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/behavior/actions/conditions/ConditionAction.java:
```diff
@@ -157,7 +157,7 @@ protected boolean condition(Actor actor) throws ClassNotFoundException, NoSuchFi
                                     break;
                                 default:
-                                    logger.error("Unsupported operation for strings: {}", tokens[2]);
+                                    logger.atError().addArgument(() -> tokens[2]).log("Unsupported operation for strings: {}");
```
same here
```suggestion
                                    logger.error("Unsupported operation for strings: {}", tokens[2]); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/behavior/actions/conditions/ConditionAction.java:
```diff
@@ -190,7 +190,8 @@ protected boolean condition(Actor actor) throws ClassNotFoundException, NoSuchFi
                                     default:
-                                        logger.error("Unknown field type or operation: {} {}", fieldValue.getClass(), tokens[2]);
+                                        logger.atError().addArgument(() -> fieldValue.getClass()).addArgument(() -> tokens[2]).
+                                                log("Unknown field type or operation: {} {}");
```
```suggestion
                                        logger.error("Unknown field type or operation: {} {}", fieldValue.getClass(), tokens[2]); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/behavior/core/ActionNode.java:
```diff
@@ -56,7 +56,8 @@ public void construct(Actor actor) {
             } catch (Exception e) {
-                logger.debug("Exception while running construct() of action {} from entity {}: ", action, actor.getEntity(), e);
+                logger.atDebug().addArgument(() -> action).addArgument(() -> actor.getEntity()).addArgument(() -> e).
+                        log("Exception while running construct() of action {} from entity {}: ");
```
```suggestion
                logger.debug("Exception while running construct() of action {} from entity {}: ", action, actor.getEntity(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/behavior/core/ActionNode.java:
```diff
@@ -67,7 +68,8 @@ public BehaviorState execute(Actor actor) {
             } catch (Exception e) {
-                logger.debug("Exception while running action {} from entity {}: ", action, actor.getEntity(), e);
+                logger.atDebug().addArgument(() -> action).addArgument(() -> actor.getEntity()).addArgument(() -> e).
+                        log("Exception while running action {} from entity {}: ");
```
```suggestion
                logger.debug("Exception while running action {} from entity {}: ", action, actor.getEntity(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/behavior/core/DecoratorNode.java:
```diff
@@ -48,7 +48,8 @@ public void construct(Actor actor) {
             } catch (Exception e) {
-                logger.info("Exception while running construct() of action {} from entity {}:", action, actor.getEntity());
+                logger.atInfo().addArgument(() -> action).addArgument(() -> actor.getEntity()).
+                        log("Exception while running construct() of action {} from entity {}:");
```
```suggestion
                logger.info("Exception while running construct() of action {} from entity {}:", action, actor.getEntity()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/behavior/core/DecoratorNode.java:
```diff
@@ -80,7 +81,8 @@ public BehaviorState execute(Actor actor) {
         } catch (Exception e) {
-            logger.info("Exception while running action {} from entity {}: {}", action, actor.getEntity(), e.getStackTrace());
+            logger.atInfo().addArgument(() -> action).addArgument(() -> actor.getEntity()).addArgument(() -> e.getStackTrace()).
+                    log("Exception while running action {} from entity {}: {}");
```
```suggestion
            logger.info("Exception while running action {} from entity {}: {}", action, actor.getEntity(), e.getStackTrace()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/console/commandSystem/MethodCommand.java:
```diff
@@ -74,20 +74,23 @@ public static void registerAvailable(Object provider, Console console, Context c
+                        addArgument(() -> method.getName()).
+                        addArgument(() -> method.getDeclaringClass().getCanonicalName()).
+                        addArgument(() -> t).
+                        log("Failed to load command method {} in class {}");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/ServerImpl.java:
```diff
@@ -346,7 +349,7 @@ private void processBlockRegistrations(NetData.NetMessage message) {
                     blockManager.receiveFamilyRegistration(family, registrationMap);
                 } catch (BlockUriParseException e) {
-                    logger.error("Received invalid block uri {}", blockFamily.getBlockUri(0));
+                    logger.atError().addArgument(() -> blockFamily.getBlockUri(0)).log("Received invalid block uri {}");
```
```suggestion
                    logger.error("Received invalid block uri {}", blockFamily.getBlockUri(0)); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/internal/GlobalStoreLoader.java:
```diff
@@ -93,7 +93,7 @@ private Prefab loadPrefab(EntityData.Prefab prefabData, Map<String, EntityData.P
                 return createPrefab(prefabData);
             } catch (Exception e) {
-                logger.error("Failed to load prefab {}", prefabData.getParentName(), e);
+                logger.atError().addArgument(() -> prefabData.getParentName()).addArgument(e).log("Failed to load prefab {}");
```
```suggestion
                logger.error("Failed to load prefab {}", prefabData.getParentName(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/serializers/WorldSerializerImpl.java:
```diff
@@ -157,7 +157,7 @@ private void createPrefab(EntityData.Prefab prefabData) {
             Assets.generateAsset(new ResourceUrn(prefabData.getName()), protoPrefab, Prefab.class);
         } catch (Exception e) {
-            logger.error("Failed to create prefab {}", prefabData.getName());
+            logger.atError().addArgument(() -> prefabData.getName()).log("Failed to create prefab {}");
```
```suggestion
            logger.error("Failed to create prefab {}", prefabData.getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/ShaderManagerLwjgl.java:
```diff
@@ -49,11 +49,11 @@ public void initShaders() {
+        logger.atInfo().addArgument(() -> GL11.glGetString(GL11.GL_VENDOR)).log("GL_VENDOR: {}");
+        logger.atInfo().addArgument(() -> GL11.glGetString(GL11.GL_RENDERER)).log("GL_RENDERER: {}");
+        logger.atInfo().addArgument(() -> GL11.glGetString(GL11.GL_VERSION)).log("GL_VERSION: {}");
+        logger.atInfo().addArgument(() -> GL11.glGetString(GL20.GL_SHADING_LANGUAGE_VERSION)).log("SHADING_LANGUAGE VERSION: {}");
```
Methods only evaluated if `logCapabilities` is evaluated whose purpose is to log.
Ignore using annotation:
```java
@SuppressWarnings("PMD.GuardLogStatement")
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/asset/UIFormat.java:
```diff
@@ -218,7 +218,8 @@ public UIWidget deserialize(JsonElement json, Type typeOfT, JsonDeserializationC
                     } catch (RuntimeException e) {
-                        logger.error("Failed to deserialize field {} of {}", field.getName(), type, e);
+                        logger.atError().addArgument(() -> field.getName()).addArgument(type).addArgument(e).
+                                log("Failed to deserialize field {} of {}");
```
```suggestion
                        logger.error("Failed to deserialize field {} of {}", field.getName(), type, e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/widgets/UIButtonWebBrowser.java:
```diff
@@ -68,7 +68,7 @@ public class UIButtonWebBrowser extends UIButton {
                 desktop.browse(new URI(this.url));
             } catch (IOException | URISyntaxException e) {
-                logger.warn("Can't open {} in default browser of your system.", this.url);
+                logger.atWarn().addArgument(() -> this.url).log("Can't open {} in default browser of your system.");
```
```suggestion
                logger.warn("Can't open {} in default browser of your system.", this.url); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/widgets/UIButtonWebBrowser.java:
```diff
@@ -83,7 +83,7 @@ public class UIButtonWebBrowser extends UIButton {
                 }
             } catch (IOException e) {
-                logger.warn("Can't recognize your OS and open the url {}.", this.url);
+                logger.atWarn().addArgument(() -> this.url).log("Can't recognize your OS and open the url {}.");
```
```suggestion
                logger.warn("Can't recognize your OS and open the url {}.", this.url); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/widgets/UIButtonWebBrowser.java:
```diff
@@ -125,7 +125,7 @@ private void showConfirmationPopup() {
             }
         } catch (MalformedURLException e) {
-            logger.error("{} is malformed", this.url, e);
+            logger.atError().addArgument(() -> this.url).addArgument(e).log("{} is malformed");
```
```suggestion
            logger.error("{} is malformed", this.url, e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/opengl/GLSLMaterial.java:
```diff
@@ -156,7 +156,7 @@ public final void doReload(MaterialData data) {
             });
         } catch (InterruptedException e) {
-            logger.error("Failed to reload {}", getUrn(), e);
+            logger.atError().addArgument(() -> getUrn()).addArgument(e).log("Failed to reload {}");
```
```suggestion
            logger.error("Failed to reload {}", getUrn(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/opengl/GLSLShader.java:
```diff
@@ -376,11 +377,11 @@ protected void doReload(ShaderData data) {
                     registerAllShaderPermutations();
                 } catch (RuntimeException e) {
-                    logger.warn("{}", e.getMessage());
+                    logger.atWarn().addArgument(() -> e.getMessage()).log("{}");
```
```suggestion
                    logger.warn("{}", e.getMessage()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/opengl/GLSLShader.java:
```diff
@@ -376,11 +377,11 @@ protected void doReload(ShaderData data) {
             });
         } catch (InterruptedException e) {
-            logger.error("Failed to reload {}", getUrn(), e);
+            logger.atError().addArgument(() -> getUrn()).addArgument(e).log("Failed to reload {}");
```
```suggestion
            logger.error("Failed to reload {}", getUrn(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java:
```diff
@@ -54,7 +54,7 @@ protected void doReload(MeshData newData) {
             GameThread.synch(() -> buildMesh(newData));
         } catch (InterruptedException e) {
-            logger.error("Failed to reload {}", getUrn(), e);
+            logger.atError().addArgument(() -> getUrn()).addArgument(e).log("Failed to reload {}");
```
```suggestion
            logger.error("Failed to reload {}", getUrn(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLSkeletalMesh.java:
```diff
@@ -112,7 +112,7 @@ protected void doReload(SkeletalMeshData newData) {
             });
         } catch (InterruptedException e) {
-            logger.error("Failed to reload {}", getUrn(), e);
+            logger.atError().addArgument(() -> getUrn()).addArgument(e).log("Failed to reload {}");
```
```suggestion
            logger.error("Failed to reload {}", getUrn(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/utilities/concurrency/TaskProcessor.java:
```diff
@@ -42,7 +42,7 @@ public void run() {
             } catch (RuntimeException e) {
                 ThreadMonitor.addError(e);
-                logger.error("Error in thread {}", Thread.currentThread().getName(), e);
+                logger.atError().addArgument(Thread.currentThread().getName()).addArgument(e).log("Error in thread {}");
```
```suggestion
                logger.error("Error in thread {}", Thread.currentThread().getName(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/block/family/BlockFamilyLibrary.java:
```diff
@@ -96,7 +96,7 @@ public <T> Optional<T> get(Class<T> type) {
             return result;
         } catch (Exception e) {
-            logger.error("Failed to load blockFamily {}", blockFamily, e);
+            logger.atError().addArgument(blockFamily).addArgument(e).log("Failed to load blockFamily {}");
```
```suggestion
            logger.error("Failed to load blockFamily {}", blockFamily, e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/chunks/localChunkProvider/LocalChunkProvider.java:
```diff
@@ -289,7 +289,7 @@ private boolean unloadChunkInternal(Vector3ic pos) {
             unloadRequestTaskMaster.put(new ChunkUnloadRequest(chunk, this));
         } catch (InterruptedException e) {
-            logger.error("Failed to enqueue unload request for {}", chunk.getPosition(), e);
+            logger.atError().addArgument(() -> chunk.getPosition()).addArgument(e).log("Failed to enqueue unload request for {}");
```
```suggestion
            logger.error("Failed to enqueue unload request for {}", chunk.getPosition(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/chunks/localChunkProvider/LocalChunkProvider.java:
```diff
@@ -299,7 +299,7 @@ void gatherBlockPositionsForDeactivate(Chunk chunk) {
             deactivateBlocksQueue.put(createBatchBlockEventMappings(chunk));
         } catch (InterruptedException e) {
-            logger.error("Failed to queue deactivation of blocks for {}", chunk.getPosition());
+            logger.atError().addArgument(() -> chunk.getPosition()).log("Failed to queue deactivation of blocks for {}");
```
```suggestion
            logger.error("Failed to queue deactivation of blocks for {}", chunk.getPosition()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/generator/internal/WorldGeneratorManager.java:
```diff
@@ -70,7 +70,8 @@ public void refresh() {
                 } catch (Exception e) {
-                    logger.error("Error loading world generator in module {}, skipping", module.getId(), e);
+                    logger.atError().addArgument(() -> module.getId()).addArgument(e).
+                            log("Error loading world generator in module {}, skipping");
```
```suggestion
                    logger.error("Error loading world generator in module {}, skipping", module.getId(), e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/viewer/layers/FacetLayers.java:
```diff
@@ -78,7 +78,7 @@ private static Collection<FacetLayer> createLayersFor(Class<? extends WorldFacet
 
         if (result.isEmpty()) {
-            logger.warn("No layers found for facet {}", facetClass.getName());
+            logger.atWarn().addArgument(() -> facetClass.getName()).log("No layers found for facet {}");
```
Simple getter for member access, neglectable impact, ignore PMD warning
```suggestion
            logger.warn("No layers found for facet {}", facetClass.getName());  //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/generation/WorldBuilder.java:
```diff
@@ -184,7 +184,8 @@ private ListMultimap<Class<? extends WorldFacet>, FacetProvider> determineProvid
                     if (!facets.contains(facet.value())) {
-                        logger.error("Facet provider for {} is missing. It is required by {}", facet.value(), provider.getClass());
+                        logger.atError().addArgument(() -> facet.value()).addArgument(() -> provider.getClass()).
+                                log("Facet provider for {} is missing. It is required by {}");
```
Both, `facet.value()` and `provider.getClass()` are already computed before the log call.
Please create and reference local variables instead.

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/generation/FacetedWorldConfigurator.java:
```diff
@@ -24,7 +24,7 @@ public FacetedWorldConfigurator(List<ConfigurableFacetProvider> providersList) {
             Component old = properties.put(provider.getConfigurationName(), provider.getConfiguration());
             if (old != null) {
-                logger.warn("Duplicate property key: {}", provider.getConfigurationName());
+                logger.atWarn().addArgument(() -> provider.getConfigurationName()).log("Duplicate property key: {}");
```
`provider.getConfigurationName()` is computed before the log call.
Please create and reference a local variable instead.

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/chunks/pipeline/ChunkProcessingPipeline.java:
```diff
@@ -115,7 +115,8 @@ private void onStageDone(PositionFuture<Chunk> future, ChunkProcessingInfo chunk
                             : chunkProcessingInfo.getChunkTaskProvider().getName();
-            logger.error("ChunkTask at position {} and stage [{}] catch error: ", chunkProcessingInfo.getPosition(), stageName, e);
+            logger.atError().addArgument(chunkProcessingInfo.getPosition()).addArgument(stageName).addArgument(e.getMessage()).
+                    log("ChunkTask at position {} and stage [{}] catch error: {}");
```
```suggestion
logger.error("ChunkTask at position {} and stage [{}] catch error: ", chunkProcessingInfo.getPosition(), stageName, e); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/block/family/BlockFamilyLibrary.java:
```diff
@@ -37,7 +37,7 @@ public BlockFamilyLibrary(ModuleEnvironment moduleEnvironment, Context context)
 
             if (!BlockFamily.class.isAssignableFrom(entry)) {
-                logger.error("Cannot load {}, must be a subclass of BlockFamily", entry.getSimpleName());
+                logger.atError().addArgument(() -> entry.getSimpleName()).log("Cannot load {}, must be a subclass of BlockFamily");
```
Only evaluated in unexpected cases during construction of `BlockFamilyLibrary` so performance impact acceptable.
```suggestion
                logger.error("Cannot load {}, must be a subclass of BlockFamily", entry.getSimpleName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java:
```diff
@@ -86,7 +86,7 @@ public void render() {
             GL30.glBindVertexArray(0);
         } else {
-            logger.error("Attempted to render disposed mesh: {}", getUrn());
+            logger.atError().addArgument(() -> getUrn()).log("Attempted to render disposed mesh: {}");
```
Only evaluated in unexpected cases. Evaluated method is simple getter for member access, so impact neglectable.
```suggestion
            logger.error("Attempted to render disposed mesh: {}", getUrn()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/opengl/GLSLMaterial.java:
```diff
@@ -228,7 +228,8 @@ public void activateFeature(ShaderProgramFeature feature) {
         } else {
-            logger.error("Attempt to activate unsupported feature {} for material {} using shader {}", feature, getUrn(), shader.getUrn());
+            logger.atError().addArgument(feature).addArgument(() -> getUrn()).addArgument(() -> shader.getUrn()).
+                    log("Attempt to activate unsupported feature {} for material {} using shader {}", feature, getUrn(), shader.getUrn());
```
Only evaluated in unexpected cases. Evaluated methods are simple getters for member access so impact neglectable.
```suggestion
            logger.error("Attempt to activate unsupported feature {} for material {} using shader {}", feature, getUrn(), shader.getUrn()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/skin/UISkinFormat.java:
```diff
@@ -145,7 +145,8 @@ public void apply(UISkinBuilder builder) {
                         entry.getValue().apply(builder);
                     } else {
-                        logger.warn("Failed to resolve UIWidget class {}, skipping style information", entry.getKey());
+                        logger.atWarn().addArgument(() -> entry.getKey()).
+                                log("Failed to resolve UIWidget class {}, skipping style information");
```
`entry.getKey()` is calculated already before the logger call.
Please create and reference a local variable.

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/videoSettings/VideoSettingsScreen.java:
```diff
@@ -328,7 +328,7 @@ public boolean onKeyEvent(NUIKeyEvent event) {
 
     public void saveSettings() {
-        logger.info("Video Settings: {}", config.renderConfigAsJson(config.getRendering()));
+        logger.atInfo().addArgument(() -> config.renderConfigAsJson(config.getRendering())).log("Video Settings: {}");
```
Only evaluated if user closes video settings screen, so impact acceptable.
```suggestion
        logger.info("Video Settings: {}", config.renderConfigAsJson(config.getRendering())); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/editor/utils/NUIEditorMenuTreeBuilder.java:
```diff
@@ -259,7 +259,7 @@ private void populateContextMenu(JsonTree node, MenuTree addTree, boolean isSkin
             }
         } else {
-            logger.warn("Could not get class for node {}", node.getValue());
+            logger.atWarn().addArgument(() -> node.getValue()).log("Could not get class for node {}");
```
Only evaluated in unexpected cases, so performance impact neglectable.
```suggestion
            logger.warn("Could not get class for node {}", node.getValue()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/asset/UIFormat.java:
```diff
@@ -175,7 +175,7 @@ public UIWidget deserialize(JsonElement json, Type typeOfT, JsonDeserializationC
                 FieldMetadata<?, ?> fieldMetadata = elementMetadata.getField(ID_FIELD);
                 if (fieldMetadata == null) {
-                    logger.warn("UIWidget type {} lacks id field", elementMetadata.getId());
+                    logger.atWarn().addArgument(() -> elementMetadata.getId()).log("UIWidget type {} lacks id field");
```
Simple getter for member access.
```suggestion
                    logger.warn("UIWidget type {} lacks id field", elementMetadata.getId()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/logic/SkeletonRenderer.java:
```diff
@@ -307,7 +307,7 @@ public void renderOpaque() {
                     boneTransforms[bone.getIndex()] = boneTransform.transpose();
                 } else {
-                    logger.warn("Unable to resolve bone \"{}\"", bone.getName());
+                    logger.atWarn().addArgument(() -> bone.getName()).log("Unable to resolve bone \"{}\"");
```
Simple getter for member access.
```suggestion
                    logger.warn("Unable to resolve bone \"{}\"", bone.getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/logic/FloatingTextRenderer.java:
```diff
@@ -77,7 +77,7 @@ private void render(Iterable<EntityRef> floatingTextEntities) {
 
             if (location == null) {
-                logger.warn("location component is not defined can't render text: {}", floatingText.text);
+                logger.atWarn().addArgument(() -> floatingText.text).log("location component is not defined can't render text: {}");
```
Simple member access
```suggestion
                logger.warn("location component is not defined can't render text: {}", floatingText.text); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/internal/GlobalStoreLoader.java:
```diff
@@ -112,7 +112,7 @@ private void loadComponentMapping(EntityData.GlobalStore globalStore) {
                 componentIdTable.put(componentMetadata.getType(), index);
             } else {
-                logger.warn("Unable to resolve component '{}'", globalStore.getComponentClass(index));
+                logger.atWarn().addArgument(globalStore.getComponentClass(index)).log("Unable to resolve component '{}'");
```
`globalStore.getComponentClass(index)` is evaluated before the logger call
Please create and reference a local variable.

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSoundSystem.java:
```diff
@@ -60,7 +60,7 @@ public void onFootstep(FootstepEvent event, EntityRef entity, LocationComponent
         if (block != null) {
             if (block.getSounds() == null) {
-                logger.error("Block '{}' has no sounds", block.getURI());
+                logger.atError().addArgument(() -> block.getURI()).log("Block '{}' has no sounds");
```
Simple getter for member access
```suggestion
                logger.error("Block '{}' has no sounds", block.getURI()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/entitySystem/metadata/EventMetadata.java:
```diff
@@ -38,7 +38,8 @@ public EventMetadata(Class<T> simpleClass, CopyStrategyLibrary copyStrategies, R
         }
         if (networkEventType != NetworkEventType.NONE && !isConstructable() && !Modifier.isAbstract(simpleClass.getModifiers())) {
-            logger.error("Event '{}' is a network event but lacks a default constructor - will not be replicated", this);
+            logger.atError().addArgument(() -> this).
+                    log("Event '{}' is a network event but lacks a default constructor - will not be replicated");
```
Simple reference of current object.
```suggestion
            logger.error("Event '{}' is a network event but lacks a default constructor - will not be replicated", this); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/subsystem/common/ConfigurationSubsystem.java:
```diff
@@ -66,7 +66,7 @@ public void preInitialise(Context rootContext) {
 
         // TODO: Move to display subsystem
-        logger.info("Video Settings: {}", config.renderConfigAsJson(config.getRendering()));
+        logger.atInfo().addArgument(() -> config.renderConfigAsJson(config.getRendering())).log("Video Settings: {}");
```
Only evaluated once before initialization.
```suggestion
        logger.info("Video Settings: {}", config.renderConfigAsJson(config.getRendering())); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/module/ModuleInstaller.java:
```diff
@@ -38,7 +38,7 @@ public class ModuleInstaller implements Callable<List<Module>> {
     public List<Module> call() throws Exception {
         Map<URI, Path> filesToDownload = getDownloadUrls(moduleList);
-        logger.info("Started downloading {} modules", filesToDownload.size());
+        logger.atInfo().addArgument(() -> filesToDownload.size()).log("Started downloading {} modules");
```
Simple map size calculation once per module download.
```suggestion
        logger.info("Started downloading {} modules", filesToDownload.size()); //NOPMD
```

jdrueckert reviewed (changes requested):
Note: only reviewed roughly half the files so far. Need to get back to the rest.
Feel free to already apply the requested changes meanwhile.

General assumption: cases in which we want to disable logging completely are extremely rare, so we mainly want to avoid method evaluation in debug logs considering that the typical default log level will be info resulting in error, warning, and info logs being written.

Cases where we want to ignore the PMD warnings instead of using Fluent API to avoid additional verbosity leading to code being harder to read:
- logs in catch clauses -> limited method evaluation
- logs in dedicated warning/error/log methods -> intention of these methods is to write logs; if at all, calling these methods should be guarded
- logs with simple getter methods

soloturn committed:
review feedback pr

introduce local varaible where necessary, otherwise ignore warning
for the sake of shorter log message line.

soloturn committed:
feedback fluent-logger pr. except to make log line shorter.

Co-authored-by: jdrueckert <jd.rueckert@googlemail.com>

jdrueckert commented on engine/src/main/java/org/terasology/engine/audio/openAL/OpenALManager.java:
```diff
@@ -64,15 +65,16 @@ public OpenALManager(AudioConfig config) throws OpenALException {
+        logger.atInfo().addArgument(() -> ALC10.alcGetString(device, ALC10.ALC_EXTENSIONS)).log("Available ALC extensions: {}");
+        logger.atInfo().addArgument(() -> ALC10.alcGetInteger(device, ALC11.ALC_MONO_SOURCES)).log("Max mono sources: {}");
+        logger.atInfo().addArgument(() -> ALC10.alcGetInteger(device, ALC11.ALC_STEREO_SOURCES)).log("Max stereo sources: {}");
+        logger.atInfo().addArgument(() -> ALC10.alcGetInteger(device, ALC10.ALC_FREQUENCY)).log("Mixer frequency: {}");
```
@soloturn you added the supression annotation, but you didn't revert the changes. please revert the changes to fluent API because they're no longer needed to fix the PMD warning now that we're suppressing it.

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/block/internal/BlockManagerImpl.java:
```diff
@@ -95,7 +95,8 @@ public void initialise(List<String> registeredBlockFamilies,
                         } else {
-                            logger.error("Missing id for block {} in provided family {}", block.getURI(), family.get().getURI());
+                            logger.atError().addArgument(() -> block.getURI()).addArgument(() -> family.get().getURI()).
+                                    log("Missing id for block {} in provided family {}");
```
`getURI()` is a simple getter method so impact is neglectable even if we call it in the worst case for every block in every block family. Also, this is an error log, so the likelihood of the method call being done "in vain" is very low.
```suggestion
                            logger.error("Missing id for block {} in provided family {}", block.getURI(), family.get().getURI()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/block/internal/BlockManagerImpl.java:
```diff
@@ -156,7 +157,8 @@ public void receiveFamilyRegistration(BlockUri familyUri, Map<String, Integer> r
                     } else {
-                        logger.error("Missing id for block {} in registered family {}", block.getURI(), familyUri);
+                        logger.atError().addArgument(() -> block.getURI()).addArgument(() -> familyUri).
+                                log("Missing id for block {} in registered family {}");
```
same reasoning as above
```suggestion
                        logger.error("Missing id for block {} in registered family {}", block.getURI(), familyUri); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/world/block/internal/BlockManagerImpl.java:
```diff
@@ -189,7 +191,7 @@ protected void registerFamily(BlockFamily family) {
     private void registerBlock(Block block, RegisteredState newState) {
         if (block.getId() != UNKNOWN_ID) {
-            logger.info("Registered Block {} with id {}", block, block.getId());
+            logger.atInfo().addArgument(() -> block).addArgument(() -> block.getId()).log("Registered Block {} with id {}");
```
Simple getter method so impact neglectable.
```suggestion
            logger.info("Registered Block {} with id {}", block, block.getId()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLTexture.java:
```diff
@@ -82,7 +82,8 @@ protected void doReload(TextureData data) {
                                         resources.id = newId;
-                                        logger.debug("Bound texture '{}' - {}", getUrn(), resources.id);
+                                        logger.atDebug().addArgument(() -> getUrn()).addArgument(resources.id).
+                                                log("Bound texture '{}' - {}");
```
Simple getter method so impact neglectable
```suggestion
                                        logger.debug("Bound texture '{}' - {}", getUrn(), resources.id); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLTexture.java:
```diff
@@ -104,7 +105,7 @@ protected void doReload(TextureData data) {
                             } else {
                                 resources.id = newId;
-                                logger.debug("Bound texture '{}' - {}", getUrn(), resources.id);
+                                logger.atDebug().addArgument(() -> getUrn()).addArgument(resources.id).log("Bound texture '{}' - {}");
```
same here
```suggestion
                                logger.debug("Bound texture '{}' - {}", getUrn(), resources.id); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/opengl/GLSLShader.java:
```diff
@@ -278,7 +278,8 @@ private void registerAllShaderPermutations() {
 
-        logger.debug("Compiled {} permutations for {}.", allPermutations.size(), getUrn());
+        logger.atDebug().addArgument(() -> allPermutations.size()).addArgument(() -> getUrn()).
+                log("Compiled {} permutations for {}.");
```
`getUrn()` is a simple getter method and `size()` complexity is probably O(1) I'd say so impact of both are neglectable
```suggestion
        logger.debug("Compiled {} permutations for {}.", allPermutations.size(), getUrn()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/opengl/GLSLShader.java:
```diff
@@ -364,7 +365,7 @@ private int compileShader(int type, Set<ShaderProgramFeature> features) {
         try {
             GameThread.synch(() -> {
-                logger.debug("Recompiling shader {}.", getUrn());
+                logger.atDebug().addArgument(() -> getUrn()).log("Recompiling shader {}.");
```
Simple getter method so impact neglectable
```suggestion
                logger.debug("Recompiling shader {}.", getUrn()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/gameDetailsScreen/GameDetailsScreen.java:
```diff
@@ -354,16 +354,19 @@ private void loadGameModules() {
                         }
-                        logger.error("Can't find any versions of module {} in your classpath!", nameVersion.getName());
+                        logger.atError().addArgument(() -> nameVersion.getName()).
+                                log("Can't find any versions of module {} in your classpath!");
```
`nameVersion.getVersion()` and `nameVersion.getName()` are called very often here in only a few lines. please introduce and reference local variables.

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/internal/NUIManagerInternal.java:
```diff
@@ -365,7 +365,8 @@ public <T extends CoreScreenLayer> T createScreen(ResourceUrn screenUri, Class<T
             } else {
-                logger.error("Screen '{}' is a '{}' and not a '{}'", screenUri, root.getClass(), expectedType);
+                logger.atError().addArgument(screenUri).addArgument(() -> root.getClass()).addArgument(expectedType).
+                        log("Screen '{}' is a '{}' and not a '{}'");
```
The `getClass()` call should be pretty cheap. also, it's only done in the rare case that the screen type doesn't match.
```suggestion
                logger.error("Screen '{}' is a '{}' and not a '{}'", screenUri, root.getClass(), expectedType); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/internal/NUIManagerInternal.java:
```diff
@@ -490,7 +491,8 @@ public <T extends ControlWidget> T addOverlay(ResourceUrn overlayUri, Class<T> e
             } else {
-                logger.error("Screen '{}' is a '{}' and not a '{}'", overlayUri, root.getClass(), expectedType);
+                logger.atError().addArgument(overlayUri).addArgument(() -> root.getClass()).addArgument(expectedType).
+                        log("Screen '{}' is a '{}' and not a '{}'");
```
same here
```suggestion
                logger.error("Screen '{}' is a '{}' and not a '{}'", overlayUri, root.getClass(), expectedType); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/config/flexible/AutoConfigTypeHandler.java:
```diff
@@ -43,7 +43,7 @@ protected PersistedData serializeNonNull(AutoConfig value, PersistedDataSerializ
                         fields.put(field.getName(), typeHandler.get().serialize(setting.get(), serializer));
                     } else {
-                        logger.error("Cannot serialize type [{}]", setting.getValueType());
+                        logger.atError().addArgument(() -> setting.getValueType()).log("Cannot serialize type [{}]");
```
Simple getter method so impact neglectable
```suggestion
                        logger.error("Cannot serialize type [{}]", setting.getValueType()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/config/flexible/AutoConfigTypeHandler.java:
```diff
@@ -77,11 +77,11 @@ public Optional<AutoConfig> deserialize(PersistedData data) {
                         }
                     } else {
-                        logger.error("Cannot deserialize type [{}]", setting.getValueType());
+                        logger.atError().addArgument(() -> setting.getValueType()).log("Cannot deserialize type [{}]");
```
same here
```suggestion
                        logger.error("Cannot deserialize type [{}]", setting.getValueType()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/config/flexible/AutoConfigTypeHandler.java:
```diff
@@ -77,11 +77,11 @@ public Optional<AutoConfig> deserialize(PersistedData data) {
                         } else {
-                            logger.error("Cannot deserialize value [{}] to type [{}]", entry.getValue(),
-                                    setting.getValueType());
+                            logger.atError().addArgument(() -> entry.getValue()).addArgument(setting.getValueType()).
+                                    log("Cannot deserialize value [{}] to type [{}]");
```
same here
```suggestion
                            logger.error("Cannot deserialize value [{}] to type [{}]", entry.getValue(),
                                    setting.getValueType()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/config/flexible/AutoConfigTypeHandler.java:
```diff
@@ -65,7 +65,7 @@ public Optional<AutoConfig> deserialize(PersistedData data) {
                 Field settingField = settingFields.get(entry.getKey());
                 if (settingField == null) {
-                    logger.warn("Cannot to find setting field with name [{}]", entry.getKey());
+                    logger.atWarn().addArgument(() -> entry.getKey()).log("Cannot to find setting field with name [{}]");
```
`getKey()` should be only O(1) so impact is neglectable
```suggestion
                    logger.warn("Cannot to find setting field with name [{}]", entry.getKey()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/config/flexible/ui/AutoConfigScreen.java:
```diff
@@ -46,7 +46,7 @@ public void initialise() {
                 mainContainer.addWidget(widget.get());
             } else {
-                logger.warn("Cannot create widget for config: {}", config.getId());
+                logger.atWarn().addArgument(() -> config.getId()).log("Cannot create widget for config: {}");
```
Simple getter method so impact neglectable
```suggestion
                logger.warn("Cannot create widget for config: {}", config.getId()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/config/flexible/ui/AutoConfigWidgetFactory.java:
```diff
@@ -79,7 +79,7 @@ public UIWidget buildWidgetFor(AutoConfig config) {
 
             if (!settingWidget.isPresent()) {
-                logger.error("Couldn't find a widget for the Setting [{}]", setting.getHumanReadableName());
+                logger.atError().addArgument(() -> setting.getHumanReadableName()).log("Couldn't find a widget for the Setting [{}]");
```
Simple getter method so impact neglectable
```suggestion
                logger.error("Couldn't find a widget for the Setting [{}]", setting.getHumanReadableName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/ComponentSystemManager.java:
```diff
@@ -68,7 +68,7 @@ public void loadSystems(ModuleEnvironment environment, NetworkMode netMode) {
         for (Class<?> type : environment.getTypesAnnotatedWith(RegisterSystem.class)) {
             if (!ComponentSystem.class.isAssignableFrom(type)) {
-                logger.error("Cannot load {}, must be a subclass of ComponentSystem", type.getSimpleName());
+                logger.atError().addArgument(type.getSimpleName()).log("Cannot load {}, must be a subclass of ComponentSystem");
```
Only called in error case and likelihood that call "in vain" is small due to error log-level
```suggestion
                logger.error("Cannot load {}, must be a subclass of ComponentSystem", type.getSimpleName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/PathManager.java:
```diff
@@ -90,7 +90,7 @@ private static Path findInstallPath() {
         Path currentDirectory = Paths.get("").toAbsolutePath();
         installationSearchPaths.add(currentDirectory);
-        LOGGER.info("PathManager: Working directory is " + currentDirectory);
+        LOGGER.atInfo().addArgument(currentDirectory).log("PathManager: Working directory is {}");
```
I don't think we need fluent API here, the following should suffice:
```suggestion
        LOGGER.info("PathManager: Working directory is {}", currentDirectory);
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/modes/StateLoading.java:
```diff
@@ -234,7 +234,7 @@ private void popStep() {
         if (!loadProcesses.isEmpty()) {
             current = loadProcesses.remove();
-            logger.debug("{}", current.getMessage());
+            logger.atDebug().addArgument(() -> current.getMessage()).log("{}");
```
In almost all cases constant string return so impact neglectable
```suggestion
            logger.debug("{}", current.getMessage()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/module/ModuleListDownloader.java:
```diff
@@ -59,13 +59,13 @@ public ModuleRegistry call() throws IOException {
 
                 ModuleMetadata meta = metaReader.read(new StringReader(json));
-                logger.debug("Read module {} - {}", meta.getId(), meta.getVersion());
+                logger.atDebug().addArgument(() -> meta.getId()).addArgument(() -> meta.getVersion()).log("Read module {} - {}");
```
Simple getter methods so impact neglectable
```suggestion
                logger.debug("Read module {} - {}", meta.getId(), meta.getVersion()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/module/ModuleListDownloader.java:
```diff
@@ -59,13 +59,13 @@ public ModuleRegistry call() throws IOException {
 
             int count = modules.size();
-            logger.info("Retrieved {} {}", count, (count == 1) ? "entry" : "entries");
+            logger.atInfo().addArgument(() -> count).addArgument(() -> (count == 1) ? "entry" : "entries").log("Retrieved {} {}");
```
simple integer comparison so impact neglectable
```suggestion
            logger.info("Retrieved {} {}", count, (count == 1) ? "entry" : "entries"); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java:
```diff
@@ -173,9 +173,10 @@ private void loadModulesFromClassPath() {
                 }
                 if (registry.add(module)) {
-                    logger.info("Loaded {} from {}", module.getId(), path);
+                    logger.atInfo().addArgument(() -> module.getId()).addArgument(() -> path).log("Loaded {} from {}");
```
simple getter method so impact neglectable
```suggestion
                    logger.info("Loaded {} from {}", module.getId(), path); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java:
```diff
@@ -173,9 +173,10 @@ private void loadModulesFromClassPath() {
+                    logger.atInfo().addArgument(() -> module.getId()).addArgument(() -> path).log("Loaded {} from {}");
                 } else {
-                    logger.info("Module {} from {} was a duplicate; not registering this copy.", module.getId(), path);
+                    logger.atInfo().addArgument(() -> module.getId()).addArgument(() -> path).
+                            log("Module {} from {} was a duplicate; not registering this copy.");
```
same here
```suggestion
                    logger.info("Module {} from {} was a duplicate; not registering this copy.", module.getId(), path); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/entitySystem/entity/internal/PojoEntityManager.java:
```diff
@@ -526,7 +526,8 @@ public <T extends Component> T addComponent(long entityId, T component) {
             notifyComponentAdded(getEntity(entityId), component.getClass());
         } else {
-            logger.error("Adding a component ({}) over an existing component for entity {}", component.getClass(), entityId);
+            logger.atError().addArgument(() -> component.getClass()).addArgument(() -> entityId).
+                    log("Adding a component ({}) over an existing component for entity {}");
```
`getClass()` should be pretty cheap so impact can be considered neglectable
```suggestion
            logger.error("Adding a component ({}) over an existing component for entity {}", component.getClass(), entityId); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/entitySystem/entity/internal/PojoEntityManager.java:
```diff
@@ -580,7 +581,8 @@ public void saveComponent(long entityId, Component component) {
 
         if (!oldComponent.isPresent()) {
-            logger.error("Saving a component ({}) that doesn't belong to this entity {}", component.getClass(), entityId);
+            logger.atError().addArgument(() -> component.getClass()).addArgument(() -> entityId).
+                    log("Saving a component ({}) that doesn't belong to this entity {}");
```
same here
```suggestion
            logger.error("Saving a component ({}) that doesn't belong to this entity {}", component.getClass(), entityId); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/game/GameManifest.java:
```diff
@@ -188,13 +188,14 @@ public void addModule(Name id, Version version) {
         var world = getWorldInfo(TerasologyConstants.MAIN_WORLD);
         if (world == null) {
-            logger.warn("{} has no MAIN_WORLD", this);
+            logger.atWarn().addArgument(() -> this).log("{} has no MAIN_WORLD");
```
why do we need fluent API here? there's no method call involved here...

jdrueckert commented on engine/src/main/java/org/terasology/engine/game/GameManifest.java:
```diff
@@ -188,13 +188,14 @@ public void addModule(Name id, Version version) {
         var generator = manager.getWorldGeneratorInfo(generatorUri);
         if (generator == null) {
-            logger.warn("{}: {} has no generator for {}", this, manager, generatorUri);
+            logger.atWarn().addArgument(() -> this).addArgument(() -> manager).addArgument(() -> generatorUri).
+                    log("{}: {} has no generator for {}");
```
same question here

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/ClientCharacterPredictionSystem.java:
```diff
@@ -84,7 +84,8 @@ public void onDestroy(final BeforeDeactivateComponent event, final EntityRef ent
     public void onCharacterStateReceived(CharacterStateEvent state, EntityRef entity) {
         if (entity.equals(localPlayer.getCharacterEntity())) {
-            logger.trace("Received new state, sequence number: {}, buffered input size {}", state.getSequenceNumber(), inputs.size());
+            logger.atTrace().addArgument(() -> state.getSequenceNumber()).addArgument(() -> inputs.size()).
+                    log("Received new state, sequence number: {}, buffered input size {}");
```
`getSequenceNumber()` is simple getter method and `size()` should be O(1) so impact neglectable
```suggestion
            logger.trace("Received new state, sequence number: {}, buffered input size {}", state.getSequenceNumber(), inputs.size()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/ClientCharacterPredictionSystem.java:
```diff
@@ -98,7 +99,7 @@ public void onCharacterStateReceived(CharacterStateEvent state, EntityRef entity
                 }
             }
-            logger.trace("Resultant input size {}", inputs.size());
+            logger.atTrace().addArgument(() -> inputs.size()).log("Resultant input size {}");
```
same here
```suggestion
            logger.trace("Resultant input size {}", inputs.size()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/console/ConsoleImpl.java:
```diff
@@ -63,13 +63,17 @@ public void registerCommand(ConsoleCommand command) {
-                    command.getSource().getClass().getCanonicalName());
+            logger.atWarn().
+                    addArgument(() -> commandName).
+                    addArgument(() -> commandRegistry.get(commandName).getSource().getClass().getCanonicalName()).
+                    addArgument(() -> command.getSource().getClass().getCanonicalName()).
+                    log("Command with name '{}' already registered by class '{}', skipping '{}'");
         } else {
             commandRegistry.put(commandName, command);
-            logger.debug("Command '{}' successfully registered for class '{}'.", commandName,
-                    command.getSource().getClass().getCanonicalName());
+            logger.atDebug().
+                    addArgument(() -> commandName).
+                    addArgument(() -> command.getSource().getClass().getCanonicalName()).
+                    log("Command '{}' successfully registered for class '{}'.");
```
revert changes and suppress PMD `GuardLogStatement` warnings via method annotation

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/console/ConsoleImpl.java:
```diff
@@ -135,7 +139,7 @@ public void addMessage(String message, MessageType type, boolean newLine) {
     public void addMessage(Message message) {
         String uncoloredText = FontUnderline.strip(FontColor.stripColor(message.getMessage()));
-        logger.info("[{}] {}", message.getType(), uncoloredText);
+        logger.atInfo().addArgument(() -> message.getType()).addArgument(() -> uncoloredText).log("[{}] {}");
```
simple getter method, so impact neglectable
```suggestion
        logger.info("[{}] {}", message.getType(), uncoloredText); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/console/commands/ServerCommands.java:
```diff
@@ -68,7 +68,7 @@ public String shutdownServer(@Sender EntityRef sender) {
         DisplayNameComponent name = clientInfo.getComponent(DisplayNameComponent.class);
 
-        logger.info("Shutdown triggered by {}", name.name);
+        logger.atInfo().addArgument(() -> name.name).log("Shutdown triggered by {}");
```
simple member access, impact neglectable
```suggestion
        logger.info("Shutdown triggered by {}", name.name); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/console/commands/ServerCommands.java:
```diff
@@ -165,7 +165,7 @@ private String kick(EntityRef clientEntity) {
             DisplayNameComponent name = clientInfo.getComponent(DisplayNameComponent.class);
 
-            logger.info("Kicking user {}", name.name);
+            logger.atInfo().addArgument(() -> name.name).log("Kicking user {}");
```
same here
```suggestion
            logger.info("Kicking user {}", name.name); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/debug/ChunkEventErrorLogger.java:
```diff
@@ -25,14 +25,14 @@ public class ChunkEventErrorLogger extends BaseComponentSystem {
     public void onNewChunk(OnChunkLoaded chunkAvailable, EntityRef worldEntity) {
         if (!loadedChunks.add(chunkAvailable.getChunkPos())) {
-            logger.error("Multiple loads of chunk {}", chunkAvailable.getChunkPos());
+            logger.atError().addArgument(() -> chunkAvailable.getChunkPos()).log("Multiple loads of chunk {}");
```
simple getter method so impact neglectable
```suggestion
            logger.error("Multiple loads of chunk {}", chunkAvailable.getChunkPos()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/debug/ChunkEventErrorLogger.java:
```diff
@@ -25,14 +25,14 @@ public class ChunkEventErrorLogger extends BaseComponentSystem {
     public void onRemoveChunk(BeforeChunkUnload chunkUnload, EntityRef worldEntity) {
         if (!loadedChunks.remove(chunkUnload.getChunkPos())) {
-            logger.error("Unload event for not loaded chunk {}", chunkUnload.getChunkPos());
+            logger.atError().addArgument(() -> chunkUnload.getChunkPos()).log("Unload event for not loaded chunk {}");
```
same here
```suggestion
            logger.error("Unload event for not loaded chunk {}", chunkUnload.getChunkPos()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/inventory/ItemPickupAuthoritySystem.java:
```diff
@@ -82,7 +82,7 @@ public void updateExtentsOnBlockItemBoxShape(OnAddedComponent event, EntityRef i
 
         if (blockFamily == null) {
-            LOGGER.warn("Prefab {} does not have a block family", itemEntity.getParentPrefab().getName());
+            LOGGER.atWarn().addArgument(() -> itemEntity.getParentPrefab().getName()).log("Prefab {} does not have a block family");
```
low impact and only called in rare case that prefab does not have a block family
```suggestion
            LOGGER.warn("Prefab {} does not have a block family", itemEntity.getParentPrefab().getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/players/PlayerSystem.java:
```diff
@@ -155,7 +155,7 @@ public void onConnect(ConnectedEvent connected, EntityRef entity) {
 
         Client clientListener = networkSystem.getOwner(entity);
-        LOGGER.info(clientListener.toString());
+        LOGGER.atInfo().addArgument(() -> clientListener.toString()).log("{}");
```
I don't think we need fluent API here. The following should suffice:
```suggestion
        LOGGER.info("{}", clientListener);
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/particles/updating/ParticleUpdaterImpl.java:
```diff
@@ -125,7 +125,8 @@ public void initialize() {
         for (Class<?> type : environment.getTypesAnnotatedWith(RegisterParticleSystemFunction.class)) {
             if (!ParticleSystemFunction.class.isAssignableFrom(type)) {
-                logger.error("Cannot register particle system function {}, must be a subclass of ParticleSystemFunction", type.getSimpleName());
+                logger.atError().addArgument(() -> type.getSimpleName()).
+                        log("Cannot register particle system function {}, must be a subclass of ParticleSystemFunction");
```
`getSimpleName()` should be O(1) so impact neglectable
```suggestion
                logger.error("Cannot register particle system function {}, must be a subclass of ParticleSystemFunction", type.getSimpleName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/serializers/EventSerializer.java:
```diff
@@ -154,7 +154,7 @@ public Class<? extends Event> getEventClass(EntityData.Event eventData) {
             }
             if (metadata == null) {
-                logger.warn("Unable to deserialize unknown event with id: {}", eventData.getType());
+                logger.atWarn().addArgument(() -> eventData.getType()).log("Unable to deserialize unknown event with id: {}");
```
`getType()` should be O(1) so impact neglectable
```suggestion
                logger.warn("Unable to deserialize unknown event with id: {}", eventData.getType()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/serializers/PrefabSerializer.java:
```diff
@@ -155,7 +155,8 @@ private void applyComponentChanges(Module context, EntityData.Prefab prefabData,
                 }
             } else if (componentData.hasType()) {
-                logger.warn("Prefab '{}' contains unknown component '{}'", prefabData.getName(), componentData.getType());
+                logger.atWarn().addArgument(() -> prefabData.getName()).addArgument(() -> componentData.getType()).
+                        log("Prefab '{}' contains unknown component '{}'");
```
`getName()` is simple getter method and `getType()` should be O(1) so impact neglectable
```suggestion
                logger.warn("Prefab '{}' contains unknown component '{}'", prefabData.getName(), componentData.getType()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/physics/bullet/BulletPhysics.java:
```diff
@@ -241,7 +241,8 @@ public HitResult rayTrace(Vector3f from, Vector3f direction, float distance, Col
                             voxelPosition);
                 } else { //we hit something we don't understand, assume its nothing and log a warning
-                    logger.warn("Unidentified object was hit in the physics engine: {}", collisionObject.userData);
+                    logger.atWarn().addArgument(() -> collisionObject.userData).
+                            log("Unidentified object was hit in the physics engine: {}");
```
simple member access so impact neglectable
```suggestion
                    logger.warn("Unidentified object was hit in the physics engine: {}", collisionObject.userData); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/physics/bullet/BulletPhysics.java:
```diff
@@ -305,7 +306,8 @@ public HitResult rayTrace(Vector3f from, Vector3f direction, float distance, Set
                                 voxelPosition);
                     } else { //we hit something we don't understand, assume its nothing and log a warning
-                        logger.warn("Unidentified object was hit in the physics engine: {}", collisionObject.userData);
+                        logger.atWarn().addArgument(() -> collisionObject.userData).
+                                log("Unidentified object was hit in the physics engine: {}");
```
same here
```suggestion
                        logger.warn("Unidentified object was hit in the physics engine: {}", collisionObject.userData); //NOPMD
```

jdrueckert reviewed (changes requested):
more review comments along the same lines as the previous

jdrueckert commented on engine/src/main/java/org/terasology/engine/entitySystem/event/internal/EventSystemImpl.java:
```diff
@@ -83,7 +83,7 @@ public void process() {
     public void registerEvent(ResourceUrn uri, Class<? extends Event> eventType) {
         eventIdMap.put(uri, eventType);
-        logger.debug("Registering event {}", eventType.getSimpleName());
+        logger.atDebug().addArgument(() -> eventType.getSimpleName()).log("Registering event {}");
```
`getSimpleName()` is most likely O(1) and only called once there when registering events, so impact can be considered neglectable
```suggestion
        logger.debug("Registering event {}", eventType.getSimpleName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/entitySystem/event/internal/EventSystemImpl.java:
```diff
@@ -95,11 +95,11 @@ public void registerEvent(ResourceUrn uri, Class<? extends Event> eventType) {
         }
 
-        logger.debug("Registering event handler {}", handlerClass.getName());
+        logger.atDebug().addArgument(() -> handlerClass.getName()).log("Registering event handler {}");
```
`getName()` is most likely O(1) and only called once there when registering event handlers, so impact can be considered neglectable
```suggestion
        logger.debug("Registering event handler {}", handlerClass.getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/entitySystem/event/internal/EventSystemImpl.java:
```diff
@@ -95,11 +95,11 @@ public void registerEvent(ResourceUrn uri, Class<? extends Event> eventType) {
         Class handlerClass = handler.getClass();
         if (!Modifier.isPublic(handlerClass.getModifiers())) {
-            logger.error("Cannot register handler {}, must be public", handlerClass.getName());
+            logger.atError().addArgument(() -> handlerClass.getName()).log("Cannot register handler {}, must be public");
```
`getName()` is most likely O(1) and only called in an unexpected case while registering event handlers, so impact is considered neglectable
```suggestion
            logger.error("Cannot register handler {}, must be public", handlerClass.getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/entitySystem/event/internal/EventSystemImpl.java:
```diff
@@ -129,16 +129,16 @@ public void registerEventHandler(ComponentSystem handler) {
                 logger.debug("Found method: {}", method);
                 if (!Event.class.isAssignableFrom(types[0]) || !EntityRef.class.isAssignableFrom(types[1])) {
-                    logger.error("Invalid event handler method: {}", method.getName());
+                    logger.atError().addArgument(() -> method.getName()).log("Invalid event handler method: {}");
```
O(1) method call, although potentially performed once per event handler method, it's only called in unexpected cases so low impact and error-level log so unlikely that call is done "in vain"
```suggestion
                    logger.error("Invalid event handler method: {}", method.getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/entitySystem/event/internal/EventSystemImpl.java:
```diff
@@ -129,16 +129,16 @@ public void registerEventHandler(ComponentSystem handler) {
-                        logger.error("Invalid event handler method: {} - {} is not a component class",
-                                method.getName(), types[i]);
+                        logger.atError().addArgument(() -> method.getName()).addArgument(types[i]).
+                                log("Invalid event handler method: {} - {} is not a component class");
```
O(1) method call, although potentially performed once per parameter of each event handler method, it's only called in unexpected cases so low impact and error-level log so unlikely that call is done "in vain"
```suggestion
                        logger.error("Invalid event handler method: {} - {} is not a component class",
                                method.getName(), types[i]); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -155,7 +155,7 @@ public String getDamageTypeName(Prefab damageType) {
         } else {
             String damageTypeName = damageType.getName();
-            logger.info("{} is missing a readable DisplayName", damageTypeName);
+            logger.atInfo().addArgument(damageTypeName).log("{} is missing a readable DisplayName");
```
why do we need fluent API here? it's just a variable reference, not a method call
```suggestion
            logger.info("{} is missing a readable DisplayName", damageTypeName);
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -308,42 +308,46 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
+                    log("Direction at client {} was different than direction at server {}");
         }
         // Assume the exact same value in case there are rounding mistakes:
         direction = event.getDirection();
```
`event.getDirection()` is used before the log, introduce a local variable and reference it in both places

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -308,42 +308,46 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
+                    addArgument(() -> getPlayerNameFromCharacter(character)).
+                    addArgument(() -> event.getOrigin()).
+                    addArgument(() -> originPos).
+                    log("Player {} seems to have cheated: It stated that it performed an action from {} but the predicted position is {}");
```
both, `event.getOrigin()` and `getPlayerNameFromCharacter(character)` are used more than once. create and reference local variables.

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -308,42 +308,46 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
-                logger.info("Denied activation attempt by {} since the used entity does not exist on the authority",
-                        getPlayerNameFromCharacter(character));
+                logger.atInfo().addArgument(() -> getPlayerNameFromCharacter(character)).
+                        log("Denied activation attempt by {} since the used entity does not exist on the authority");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -308,42 +308,46 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
-                logger.info("Denied activation attempt by {} since it does not own the entity at the authority",
-                        getPlayerNameFromCharacter(character));
+                logger.atInfo().addArgument(() -> getPlayerNameFromCharacter(character)).
+                        log("Denied activation attempt by {} since it does not own the entity at the authority");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -308,42 +308,46 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
-                logger.info("Denied activation attempt by {} since it is not properly marked as owned entity usage",
-                        getPlayerNameFromCharacter(character));
+                logger.atInfo().addArgument(() -> getPlayerNameFromCharacter(character)).
+                        log("Denied activation attempt by {} since it is not properly marked as owned entity usage");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -308,42 +308,46 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
-                logger.info("Denied activation attempt by {} since the target does not exist on the authority",
-                        getPlayerNameFromCharacter(character));
+                logger.atInfo().addArgument(() -> getPlayerNameFromCharacter(character)).
+                        log("Denied activation attempt by {} since the target does not exist on the authority");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -360,8 +364,8 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
-                logger.info("Denied activation attempt by {} since at the authority there was nothing to activate at that place",
-                        getPlayerNameFromCharacter(character));
+                logger.atInfo().addArgument(() -> getPlayerNameFromCharacter(character)).
+                        log("Denied activation attempt by {} since at the authority there was nothing to activate at that place");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -371,31 +375,31 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
-                logger.info("Denied activation attempt by {} since at the authority another entity would have been activated",
-                        getPlayerNameFromCharacter(character));
+                logger.atInfo().addArgument(() -> getPlayerNameFromCharacter(character)).
+                        log("Denied activation attempt by {} since at the authority another entity would have been activated");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -371,31 +375,31 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
-                logger.info("Denied activation attempt by {} since at the authority the object got hit at a differnt position",
-                        getPlayerNameFromCharacter(character));
+                logger.atInfo().addArgument(() -> getPlayerNameFromCharacter(character)).
+                        log("Denied activation attempt by {} since at the authority the object got hit at a differnt position");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -371,31 +375,31 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
-                logger.info("Denied activation attempt by {} since the event was not properly labeled as having a target",
-                        getPlayerNameFromCharacter(character));
+                logger.atInfo().addArgument(() -> getPlayerNameFromCharacter(character)).
+                        log("Denied activation attempt by {} since the event was not properly labeled as having a target");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -371,31 +375,31 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
-                logger.info("Denied activation attempt by {} since the event was not properly labeled as having a hit position",
-                        getPlayerNameFromCharacter(character));
+                logger.atInfo().addArgument(() -> getPlayerNameFromCharacter(character)).
+                        log("Denied activation attempt by {} since the event was not properly labeled as having a hit position");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -371,31 +375,31 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
-                logger.info("Denied activation attempt by {} since the event was not properly labeled as having a hit delta",
-                        getPlayerNameFromCharacter(character));
+                logger.atInfo().addArgument(() -> getPlayerNameFromCharacter(character)).
+                        log("Denied activation attempt by {} since the event was not properly labeled as having a hit delta");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/console/commandSystem/MethodCommand.java:
```diff
@@ -74,20 +74,23 @@ public static void registerAvailable(Object provider, Console console, Context c
-            logger.debug("Registering command method {} in class {}", method.getName(),
-                    method.getDeclaringClass().getCanonicalName());
+            logger.atDebug().addArgument(() -> method.getName()).addArgument(() -> method.getDeclaringClass().getCanonicalName()).
+                    log("Registering command method {} in class {}");
```
`provider.getClass()` is used before (line 74), create and reference local variable; ignore PMD with respect to the low-impact `getSimpleName()` call
`method.getName()` is referenced in various logs, create and reference local variable in method loop

jdrueckert commented on engine/src/main/java/org/terasology/engine/logic/console/commandSystem/MethodCommand.java:
```diff
@@ -74,20 +74,23 @@ public static void registerAvailable(Object provider, Console console, Context c
-                logger.debug("Registered command method {} in class {}", method.getName(),
-                        method.getDeclaringClass().getCanonicalName());
+                logger.atDebug().addArgument(() -> method.getName()).addArgument(() -> method.getDeclaringClass().getCanonicalName()).
+                        log("Registered command method {} in class {}");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/ClientConnectionHandler.java:
```diff
@@ -160,8 +160,12 @@ private void receiveModuleStart(ChannelHandlerContext channelHandlerContext,
+                        addArgument(() -> moduleDataHeader.getVersion()).
+                        addArgument(() -> sizeString).
+                        addArgument(() -> missingModules.size()).
+                        log("Downloading {}: {} ({}, {} modules remain)");
```
`moduleDataHeader.getId()` is already stored in local `moduleId` variable, reference that
`moduleDataHeader.getVersion()`and `missingModules.size()` are used in the previous line (`joinStatus`) as well. create and reference local variables

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/ClientConnectionHandler.java:
```diff
@@ -176,7 +180,8 @@ private void receiveModuleStart(ChannelHandlerContext channelHandlerContext,
         } else {
-            logger.error("Received unwanted module {}:{} from server", moduleDataHeader.getId(), moduleDataHeader.getVersion());
+            logger.atError().addArgument(() -> moduleDataHeader.getId()).addArgument(() -> moduleDataHeader.getVersion()).
+                    log("Received unwanted module {}:{} from server");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/NetClient.java:
```diff
@@ -363,7 +363,7 @@ public boolean isLocal() {
 
     void send(NetData.NetMessage data) {
-        logger.trace("Sending packet with size {}", data.getSerializedSize());
+        logger.atTrace().addArgument(() -> data.getSerializedSize()).log("Sending packet with size {}");
```
`data.getSerializedSize()` is used two lines later again, so create and reference a local variable

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/NetClient.java:
```diff
@@ -524,7 +525,8 @@ private void processEvents(NetData.NetMessage message) {
                     } else {
-                        logger.warn("Received event {} for non-owned entity {} from {}", event, target, this);
+                        logger.atWarn().addArgument(() -> event).addArgument(target).addArgument(() -> this).
+                                log("Received event {} for non-owned entity {} from {}");
```
why do we need fluent API here? there's no method call involved

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/NetworkSystemImpl.java:
```diff
@@ -183,7 +183,7 @@ public void host(int port, boolean dedicatedServer) throws HostingFailedExceptio
                 logger.info("Started server on port {}", port);
                 if (config.getServerMOTD() != null) {
-                    logger.info("Server MOTD is \"{}\"", config.getServerMOTD());
+                    logger.atInfo().addArgument(() -> config.getServerMOTD()).log("Server MOTD is \"{}\"");
```
either use local variable before null check or ignore PMD warning here because it's a simple getter method call

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/NetworkSystemImpl.java:
```diff
@@ -239,7 +239,7 @@ public JoinStatus join(String address, int port) throws InterruptedException {
                         }
                     } else {
-                        logger.warn("Failed to connect to server", connectCheck.cause());
+                        logger.atWarn().addArgument(connectCheck.cause()).log("Failed to connect to server");
```
`connectCheck.cause()` is called again three lines later, create and reference local variable

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/NetworkSystemImpl.java:
```diff
@@ -538,7 +538,8 @@ public void unregisterNetworkEntity(EntityRef entity) {
             if (netComponent != null) {
-                logger.debug("Unregistering network entity: {} with netId {}", entity, netComponent.getNetworkId());
+                logger.atDebug().addArgument(entity).addArgument(netComponent.getNetworkId()).
+                        log("Unregistering network entity: {} with netId {}");
```
`netComponent.getNetworkId()` is used multiple times again after logging, so create and reference local variable

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/NetworkSystemImpl.java:
```diff
@@ -826,7 +827,7 @@ private void processRemovedClient(Client client) {
         clientList.remove(client);
         clientPlayerLookup.remove(client.getEntity());
-        logger.info("Client disconnected: {}", client.getName());
+        logger.atInfo().addArgument(() -> client.getName()).log("Client disconnected: {}");
```
info-level and only called when client removed so impact neglectable.
```suggestion
        logger.info("Client disconnected: {}", client.getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/NetworkSystemImpl.java:
```diff
@@ -855,7 +856,7 @@ private void processNewClient(NetClient client) {
 
         // log after connect so that the name has been set:
-        logger.info("New client entity: {}", client.getEntity());
+        logger.atInfo().addArgument(() -> client.getEntity()).log("New client entity: {}");
```
info-level log and only called when client added so impact neglectable
```suggestion
        logger.info("New client entity: {}", client.getEntity()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/NetworkSystemImpl.java:
```diff
@@ -972,7 +973,8 @@ private <T> Map<Class<? extends T>, Integer> generateIds(ClassLibrary<T> classLi
             for (FieldMetadata<?, ?> field : metadata.getFields()) {
                 if (fieldId >= 256) {
-                    logger.error("Class {} has too many fields (>255), serialization will be incomplete", metadata.getId());
+                    logger.atError().addArgument(() -> metadata.getId()).
+                            log("Class {} has too many fields (>255), serialization will be incomplete");
```
simple getter method and only called in unexpected case once so impact neglectable
```suggestion
                    logger.error("Class {} has too many fields (>255), serialization will be incomplete", metadata.getId()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/NetworkSystemImpl.java:
```diff
@@ -1000,11 +1002,12 @@ private <T> Map<Class<? extends T>, Integer> applySerializationInfo(List<NetData
                         field.setId(info.getFieldIds().byteAt(i));
                     } else {
-                        logger.error("Server has unknown field '{}' on '{}'", info.getFieldName(i), info.getName());
+                        logger.atError().addArgument(info.getFieldName(i)).addArgument(() -> info.getName()).
+                                log("Server has unknown field '{}' on '{}'");
```
`info.getFieldName(i)` referenced before (line 1000) and `info.getName()` too (line 996), so create and reference local variables

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/NetworkSystemImpl.java:
```diff
@@ -1000,11 +1002,12 @@ private <T> Map<Class<? extends T>, Integer> applySerializationInfo(List<NetData
                 }
             } else {
-                logger.error("Server has unknown class '{}'", info.getName());
+                logger.atError().addArgument(() -> info.getName()).log("Server has unknown class '{}'");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/ServerImpl.java:
```diff
@@ -357,7 +360,8 @@ private void updateEntity(NetData.UpdateEntityMessage updateEntity) {
             NetworkComponent netComp = currentEntity.getComponent(NetworkComponent.class);
             if (netComp == null) {
-                logger.error("Updating entity with no network component: {}, expected netId {}", currentEntity, updateEntity.getNetId());
+                logger.atError().addArgument(() -> currentEntity).addArgument(() -> updateEntity.getNetId()).
+                        log("Updating entity with no network component: {}, expected netId {}");
```
`updateEntity.getNetId()` called before and later-on, so create and reference local variable

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/ServerImpl.java:
```diff
@@ -371,10 +375,14 @@ private void updateEntity(NetData.UpdateEntityMessage updateEntity) {
             }
             if (netComp.getNetworkId() != updateEntity.getNetId()) {
-                logger.error("Network ID lost in update: {}, {} -> {}", currentEntity, updateEntity.getNetId(), netComp.getNetworkId());
+                logger.atError().
+                        addArgument(() -> currentEntity).
+                        addArgument(() -> updateEntity.getNetId()).
+                        addArgument(() -> netComp.getNetworkId()).
+                        log("Network ID lost in update: {}, {} -> {}");
```
see above, same for `netComp.getNetworkId()`

jdrueckert commented on engine/src/main/java/org/terasology/engine/network/internal/ServerImpl.java:
```diff
@@ -371,10 +375,14 @@ private void updateEntity(NetData.UpdateEntityMessage updateEntity) {
             }
         } else {
-            logger.warn("Received update for non-existent entity {}", updateEntity.getNetId());
+            logger.atWarn().addArgument(() -> updateEntity.getNetId()).log("Received update for non-existent entity {}");
```
see above

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/serializers/ComponentSerializer.java:
```diff
@@ -207,7 +208,7 @@ public EntityData.Component serialize(Component component) {
         ComponentMetadata<?> componentMetadata = componentLibrary.getMetadata(component.getClass());
         if (componentMetadata == null) {
-            logger.error("Unregistered component type: {}", component.getClass());
+            logger.atError().addArgument(() -> component.getClass()).log("Unregistered component type: {}");
```
used before in line 209, create and reference local variable

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/serializers/ComponentSerializer.java:
```diff
@@ -262,7 +263,7 @@ public EntityData.Component serialize(Component base, Component delta) {
         ComponentMetadata<?> componentMetadata = componentLibrary.getMetadata(base.getClass());
         if (componentMetadata == null) {
-            logger.error("Unregistered component type: {}", base.getClass());
+            logger.atError().addArgument(() -> base.getClass()).log("Unregistered component type: {}");
```
same here

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/serializers/ComponentSerializer.java:
```diff
@@ -325,7 +326,7 @@ public ComponentMetadata<? extends Component> getComponentMetadata(EntityData.Co
             }
             if (metadata == null) {
-                logger.warn("Unable to deserialize unknown component with id: {}", componentData.getTypeIndex());
+                logger.atWarn().addArgument(() -> componentData.getTypeIndex()).log("Unable to deserialize unknown component with id: {}");
```
same here

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/serializers/ComponentSerializer.java:
```diff
@@ -337,7 +338,7 @@ public ComponentMetadata<? extends Component> getComponentMetadata(EntityData.Co
             }
             if (metadata == null) {
-                logger.warn("Unable to deserialize unknown component type: {}", componentData.getType());
+                logger.atWarn().addArgument(() -> componentData.getType()).log("Unable to deserialize unknown component type: {}");
```
same here

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/serializers/NetworkEntitySerializer.java:
```diff
@@ -130,7 +130,7 @@ private void serializeComponentDelta(Component oldComponent, Component newCompon
         ComponentMetadata<?> componentMetadata = componentLibrary.getMetadata(oldComponent.getClass());
         if (componentMetadata == null) {
-            logger.error("Unregistered component type: {}", oldComponent.getClass());
+            logger.atError().addArgument(() -> oldComponent.getClass()).log("Unregistered component type: {}");
```
`oldComponent.getClass()` used before in line 131, create and reference local variable

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/serializers/NetworkEntitySerializer.java:
```diff
@@ -165,7 +165,7 @@ private void serializeComponentFull(Component component, boolean ignoreIfNoField
         ComponentMetadata<?> componentMetadata = componentLibrary.getMetadata(component.getClass());
         if (componentMetadata == null) {
-            logger.error("Unregistered component type: {}", component.getClass());
+            logger.atError().addArgument(() -> component.getClass()).log("Unregistered component type: {}");
```
same here

jdrueckert commented on engine/src/main/java/org/terasology/engine/persistence/serializers/NetworkEntitySerializer.java:
```diff
@@ -218,7 +218,11 @@ public void deserializeOnto(MutableComponentContainer entity, EntityData.PackedE
                 ReplicatedFieldMetadata fieldMetadata = metadata.getField(fieldId);
                 if (fieldMetadata != null && fieldCheck.shouldDeserialize(metadata, fieldMetadata)) {
-                    logger.trace("Deserializing field {} of component {} as value {}", fieldMetadata, metadata, entityData.getFieldValue(fieldPos));
+                    logger.atTrace().
+                            addArgument(() -> fieldMetadata).
+                            addArgument(() -> metadata).
+                            addArgument(entityData.getFieldValue(fieldPos)).
+                            log("Deserializing field {} of component {} as value {}");
```
`entityData.getFieldValue(fieldPos)` will be called in next line anyway so create and reference local variable

jdrueckert commented on engine/src/main/java/org/terasology/engine/recording/EventSystemReplayImpl.java:
```diff
@@ -276,7 +276,7 @@ private EntityRef getEntityRef(RecordedEvent recordedEvent) {
     public void registerEvent(ResourceUrn uri, Class<? extends Event> eventType) {
         eventIdMap.put(uri, eventType);
-        logger.debug("Registering event {}", eventType.getSimpleName());
+        logger.atDebug().addArgument(() -> eventType.getSimpleName()).log("Registering event {}");
```
O(1) method call only called once when registering an event, so impact considered neglectable
```suggestion
        logger.debug("Registering event {}", eventType.getSimpleName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/recording/EventSystemReplayImpl.java:
```diff
@@ -303,11 +303,11 @@ private boolean shouldAddToLibrary(Class<? extends Event> eventType) {
         Class handlerClass = handler.getClass();
         if (!Modifier.isPublic(handlerClass.getModifiers())) {
-            logger.error("Cannot register handler {}, must be public", handlerClass.getName());
+            logger.atError().addArgument(() -> handlerClass.getName()).log("Cannot register handler {}, must be public");
```
```suggestion
            logger.error("Cannot register handler {}, must be public", handlerClass.getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/recording/EventSystemReplayImpl.java:
```diff
@@ -303,11 +303,11 @@ private boolean shouldAddToLibrary(Class<? extends Event> eventType) {
         }
 
-        logger.debug("Registering event handler {}", handlerClass.getName());
+        logger.atDebug().addArgument(() -> handlerClass.getName()).log("Registering event handler {}");
```
O(1) method call, only done once when registering event handlers
```suggestion
        logger.debug("Registering event handler {}", handlerClass.getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/recording/EventSystemReplayImpl.java:
```diff
@@ -336,15 +336,16 @@ public void registerEventHandler(ComponentSystem handler) {
                 logger.debug("Found method: {}", method);
                 if (!Event.class.isAssignableFrom(types[0]) || !EntityRef.class.isAssignableFrom(types[1])) {
-                    logger.error("Invalid event handler method: {}", method.getName());
+                    logger.atError().addArgument(() -> method.getName()).log("Invalid event handler method: {}");
```
potentially called once for each event handler method but only called in unexpected case and error-level log so unlikely that called "in vain"
```suggestion
                    logger.error("Invalid event handler method: {}", method.getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/recording/EventSystemReplayImpl.java:
```diff
@@ -336,15 +336,16 @@ public void registerEventHandler(ComponentSystem handler) {
                 for (int i = 2; i < types.length; ++i) {
                     if (!Component.class.isAssignableFrom(types[i])) {
-                        logger.error("Invalid event handler method: {} - {} is not a component class", method.getName(), types[i]);
+                        logger.atError().addArgument(() -> method.getName()).addArgument(types[i]).
+                                log("Invalid event handler method: {} - {} is not a component class");
```
potentially called once for each parameter of every event handler method but only called in unexpected case and error-level log so unlikely that called "in vain"
```suggestion
                        logger.error("Invalid event handler method: {} - {} is not a component class", method.getName(), types[i]); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/AbstractNode.java:
```diff
@@ -204,8 +204,8 @@ public boolean addOutputBufferPairConnection(int id, BufferPair bufferPair) {
             if (localBufferPairConnection.getData() != null) {
-                logger.warn("Adding output buffer pair to slot id {} of {} node overwrites data of existing connection: {}",
-                        id, this.nodeUri, localBufferPairConnection);
+                logger.atWarn().addArgument(id).addArgument(() -> this.nodeUri).addArgument(localBufferPairConnection).
+                        log("Adding output buffer pair to slot id {} of {} node overwrites data of existing connection: {}");
```
why do we need fluent API here? there's no method call involved :thinking:

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/AbstractNode.java:
```diff
@@ -241,8 +241,8 @@ public boolean addOutputBufferPairConnection(int id, BufferPairConnection from)
             if (localBufferPairConnection.getData() != null) {
-                logger.warn("Adding output buffer pair connection to slot id {} of {} node overwrites data of existing connection: {}",
-                        id, this.nodeUri, localBufferPairConnection);
+                logger.atWarn().addArgument(id).addArgument(() -> this.nodeUri).addArgument(localBufferPairConnection).
+                        log("Adding output buffer pair connection to slot id {} of {} node overwrites data of existing connection: {}");
```
same here

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/AbstractNode.java:
```diff
@@ -303,8 +303,8 @@ protected boolean addOutputFboConnection(int id, FBO fboData) {
             if (fboConnection.getData() != null) {
-                logger.warn("Adding output fbo data to slot id {} of {} node overwrites data of existing connection: {}",
-                        id, this.nodeUri, fboConnection);
+                logger.atWarn().addArgument(id).addArgument(() -> this.nodeUri).addArgument(fboConnection).
+                        log("Adding output fbo data to slot id {} of {} node overwrites data of existing connection: {}");
```
same here

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/AbstractNode.java:
```diff
@@ -443,7 +443,8 @@ public int getNumberOfOutputConnections() {
         DependencyConnection connection = inputConnections.get(name);
         if (connection == null) {
-            logger.error("Getting input connection named {} returned null. No such input connection in {}", name, this);
+            logger.atError().addArgument(name).addArgument(this).
+                    log("Getting input connection named {} returned null. No such input connection in {}");
```
same here

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/AbstractNode.java:
```diff
@@ -453,7 +454,8 @@ public DependencyConnection getInputConnection(String name) {
         if (connection == null) {
-            logger.error("Getting output connection named {} returned null. No such output connection in {}.", name, this);
+            logger.atError().addArgument(name).addArgument(this).
+                    log("Getting output connection named {} returned null. No such output connection in {}.");
```
same here

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/AbstractNode.java:
```diff
@@ -472,7 +474,7 @@ public List<String> getOutputConnectionsList() {
 
     public void disconnectInputFbo(int inputId) {
-        logger.info("Disconnecting {} input Fbo number {}", this.getUri(), inputId);
+        logger.atInfo().addArgument(() -> this.getUri()).addArgument(inputId).log("Disconnecting {} input Fbo number {}");
```
simple getter method
```suggestion
        logger.info("Disconnecting {} input Fbo number {}", this.getUri(), inputId); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/AbstractNode.java:
```diff
@@ -547,8 +549,8 @@ public void dispose() {
         if (stateChange.isTheDefaultInstance()) {
-            logger.error("Attempted to add default state change {} to the set of desired state changes. (Node: {})",
-                    stateChange.getClass().getSimpleName(), this);
+            logger.atError().addArgument(() -> stateChange.getClass().getSimpleName()).addArgument(this).
+                    log("Attempted to add default state change {} to the set of desired state changes. (Node: {})");
```
error-level log so unlikely that method called "in vain"
```suggestion
            logger.error("Attempted to add default state change {} to the set of desired state changes. (Node: {})",
                    stateChange.getClass().getSimpleName(), this); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/ModuleRendering.java:
```diff
@@ -118,9 +118,11 @@ protected FBO generateWithDimensions(FboConfig fboConfig, FBO.Dimensions dimensi
         if (fbo.getStatus() == FBO.Status.INCOMPLETE) {
-            logger.error("FBO {} is incomplete. Look earlier in the log for details.", fboConfig.getName());
+            logger.atError().addArgument(() -> fboConfig.getName()).
+                    log("FBO {} is incomplete. Look earlier in the log for details.");
         } else if (fbo.getStatus() == FBO.Status.UNEXPECTED) {
-            logger.error("FBO {} has generated an unexpected status code. Look earlier in the log for details.", fboConfig.getName());
+            logger.atError().addArgument(() -> fboConfig.getName()).
+                    log("FBO {} has generated an unexpected status code. Look earlier in the log for details.");
```
`fboConfig.getName()` is simple getter method, either suppress PMD warning with comment or introduce and reference local variable

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderGraph.java:
```diff
@@ -52,8 +52,8 @@ public void addNode(Node node) {
             Node aNode = akaNodeMap.get(nodeAka);
-            logger.info("Node {} also known as {} already matches existing node with uri {} - attempting replacing...",
-                    nodeUri, nodeAka, aNode.getUri());
+            logger.atInfo().addArgument(nodeUri).addArgument(nodeAka).addArgument(() -> aNode.getUri()).
+                    log("Node {} also known as {} already matches existing node with uri {} - attempting replacing...");
```
simple getter method
```suggestion
            logger.info("Node {} also known as {} already matches existing node with uri {} - attempting replacing...",
                    nodeUri, nodeAka, aNode.getUri()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderGraph.java:
```diff
@@ -149,7 +149,8 @@ private void connect(Node... nodeList) {
                     graph.putEdge(fromNode, toNode);
                 } else {
-                    logger.warn("Trying to connect two already connected nodes, {} and {}", fromNode.getUri(), toNode.getUri());
+                    logger.atWarn().addArgument(fromNode.getUri()).addArgument(toNode.getUri()).
+                            log("Trying to connect two already connected nodes, {} and {}", fromNode.getUri(), toNode.getUri());
```
simple getter methods
```suggestion
                    logger.warn("Trying to connect two already connected nodes, {} and {}", fromNode.getUri(), toNode.getUri()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderGraph.java:
```diff
@@ -169,7 +170,8 @@ public void disconnect(Node fromNode, Node toNode) {
 
         if (!graph.hasEdgeConnecting(fromNode, toNode)) {
-            logger.warn("Trying to disconnect two nodes that aren't connected, {} and {}", fromNode.getUri(), toNode.getUri());
+            logger.atWarn().addArgument(() -> fromNode.getUri()).addArgument(() -> toNode.getUri()).
+                    log("Trying to disconnect two nodes that aren't connected, {} and {}");
```
same here
```suggestion
            logger.warn("Trying to disconnect two nodes that aren't connected, {} and {}", fromNode.getUri(), toNode.getUri()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderGraph.java:
```diff
@@ -275,8 +277,8 @@ private void connectFbo(Node toNode, int inputFboId, DependencyConnection fromCo
             //TODO update
-            logger.info("{}.connectFbo({}, {}): Connection already existed. Testing for its connections..",
-                    toNode.getUri(), inputFboId, fromConnection.getName());
+            logger.atInfo().addArgument(() -> toNode.getUri()).addArgument(inputFboId).addArgument(() -> fromConnection.getName()).
+                    log("{}.connectFbo({}, {}): Connection already existed. Testing for its connections..");
```
simple getter methods
```suggestion
            logger.info("{}.connectFbo({}, {}): Connection already existed. Testing for its connections..",
                    toNode.getUri(), inputFboId, fromConnection.getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderGraph.java:
```diff
@@ -339,8 +342,8 @@ private void connectBufferPair(Node toNode, int inputConnectionId, DependencyCon
             //TODO update
-            logger.info("{}.connectFbo({}, {}): Connection already existed. Testing for its connections..",
-                    toNode.getUri(), inputConnectionId, fromConnection.getName());
+            logger.atInfo().addArgument(() -> toNode.getUri()).addArgument(inputConnectionId).addArgument(() -> fromConnection.getName()).
+                    log("{}.connectFbo({}, {}): Connection already existed. Testing for its connections..");
```
simple getter methods
```suggestion
            logger.info("{}.connectFbo({}, {}): Connection already existed. Testing for its connections..",
                    toNode.getUri(), inputConnectionId, fromConnection.getName()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderGraph.java:
```diff
@@ -529,13 +533,14 @@ public void reconnectInputToOutput(String fromNodeUri, int outputId, Node toNode
     private void reconnectInputToOutput(Node toNode, int inputId, DependencyConnection fromConnection,
                                         ConnectionType connectionType, boolean disconnectPrevious) {
-        logger.debug("Attempting reconnection of {} to {}'s output.", toNode.getUri(), fromConnection.getParentNode());
+        logger.atDebug().addArgument(() -> toNode.getUri()).addArgument(() -> fromConnection.getParentNode()).
+                log("Attempting reconnection of {} to {}'s output.");
```
simple getter methods
```suggestion
        logger.debug("Attempting reconnection of {} to {}'s output.", toNode.getUri(), fromConnection.getParentNode()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderGraph.java:
```diff
@@ -529,13 +533,14 @@ public void reconnectInputToOutput(String fromNodeUri, int outputId, Node toNode
         if (!fromConnection.getConnectedConnections().isEmpty()) {
-            logger.warn("WARNING: destination connection ({}) is already connected to ({})",
-                    fromConnection, fromConnection.getConnectedConnections());
+            logger.atWarn().addArgument(fromConnection).addArgument(() -> fromConnection.getConnectedConnections()).
+                    log("WARNING: destination connection ({}) is already connected to ({})");
```
simple getter method
```suggestion
            logger.warn("WARNING: destination connection ({}) is already connected to ({})",
                    fromConnection, fromConnection.getConnectedConnections()); //NOPMD
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderTaskListGenerator.java:
```diff
@@ -48,15 +48,15 @@ private void logIntermediateRendererListForDebugging(List<Node> orderedNodes) {
-                logger.info("----- {}", node.getClass().getSimpleName());
+                logger.atInfo().addArgument(() -> node.getClass().getSimpleName()).log("----- {}");
 
                 // printing out individual desired state changes
                 for (StateChange desiredStateChange : node.getDesiredStateChanges()) {
-                    logger.info("{}", desiredStateChange);
+                    logger.atInfo().addArgument(desiredStateChange).log("{}");
                 }
 
                 // printing out process() statement
-                logger.info("{}: process()", node);
+                logger.atInfo().addArgument(node).log("{}: process()");
```
dedicated log method, suppress `GuardLogStatement` PMD warnings via method annotation

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderTaskListGenerator.java:
```diff
@@ -167,23 +167,28 @@ public List<RenderPipelineTask> generateFrom(List<Node> orderedNodes) {
-            logger.debug("===== RENDERER TASK LIST =================================");
+            logger.atDebug().log("===== RENDERER TASK LIST =================================");
             logList(taskList);
-            logger.debug("----------------------------------------------------------");
-            logger.debug("Task list generated in {} ms", String.format("%.3f", (endTimeInNanoSeconds - startTimeInNanoSeconds) / 1000000f));
-            logger.debug("{} nodes, {} enabled - {} tasks (excluding marker tasks) out of {} potential tasks.",
-                    nodeList.size(), enabledNodes, taskList.size() - enabledNodes, potentialTasks);
-            logger.debug("----------------------------------------------------------");
+            logger.atDebug().log("----------------------------------------------------------");
+            logger.atDebug().addArgument(() -> String.format("%.3f", (endTimeInNanoSeconds - startTimeInNanoSeconds) / 1000000f)).
+                    log("Task list generated in {} ms");
+            logger.atDebug().
+                    addArgument(() -> nodeList.size()).
+                    addArgument(enabledNodes).
+                    addArgument(taskList.size() - enabledNodes).
+                    addArgument(potentialTasks).
+                    log("{} nodes, {} enabled - {} tasks (excluding marker tasks) out of {} potential tasks.");
+            logger.atDebug().log("----------------------------------------------------------");
         // }
```
utilize log guard (see outcommented code)

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderTaskListGenerator.java:
```diff
@@ -167,23 +167,28 @@ public List<RenderPipelineTask> generateFrom(List<Node> orderedNodes) {
     private void logList(List<?> list) {
         for (Object object : list) {
-            logger.debug("{}", object);
+            logger.atDebug().addArgument(object).log("{}");
```
why do we need fluent API here when there's no method call involved...?

soloturn replied:
what do you mean by "need". fluent api is not painful, and not discouraged :) additionally, this is a loop over a list.

jdrueckert replied:
It's slightly more painful to read, so I would slightly discourage using it except for cases where the added verbosity makes sense due to some kind of need. if there's no need I'd always go for the easier to read version.

soloturn replied:
easier to read is in the eye of the one reading. for me fluent is now easier to read :) but i understand, having not written fluent for you it looks stranger. but here, the main reason is that it loops over a list of objects.

jdrueckert replied:
it may loop over a list of objects, but there's no additional calculation involved that is performed for every object...
did PMD complain about this?

jdrueckert reviewed (dismissed):
Finally done with all files. Please re-request a review once you implemented all requested changes :slightly_smiling_face:

soloturn committed:
Update engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLTexture.java

Co-authored-by: jdrueckert <jd.rueckert@googlemail.com>

soloturn committed:
line length warning removed

jdrueckert committed:
refactor: use less verbose varient of fluent API

jdrueckert committed:
fix: //NOPMD doesn't work if log spans multiple lines

- removed line break or switched back to fluent API (using less verbose variant though)

jdrueckert committed:
refactor: introduce and reference local variables

jdrueckert committed:
refactor: suppress guardlogstatement warnings for methods intended for logging

jdrueckert committed:
revert: unnecessary usage of fluent API

jdrueckert committed:
qa: address remaining PMD guardlogstatement warnings

jdrueckert committed:
Merge branch 'develop' into qa/fluent-logger

jdrueckert committed:
fix: log format

jdrueckert committed:
Merge remote-tracking branch 'origin/qa/fluent-logger' into qa/fluent-logger

jdrueckert reviewed (approved):
@soloturn I went through all files again, applied the requested changes that you resolved without actually applying them. I also refactored your changes to use the less verbose variant of the fluent API that was discussed a while ago in our group sync.

From my pov this is good to go now, so I'm approving it.

For future reference two comments:
1. Please try avoid force pushes as they can remove review comments. Force pushes are rarely necessary, usually merging upstream changes or the most recent main branch state in should be sufficient.
2. Code refactorings (please don't confuse them with functional changes: introducing a local variable and referencing it does not modify the code's functionality) are fine when addressing code scanner findings.

While writing this second point and remembering your last comment on this PR, I also want to share an observation:
Based on the discussions we had during the work on our last milestone, I expected this PR to be about addressing code scanner findings (in particular PMD GuardLogStatement warnings).
According to your comment, in which you state that this PR is about "writing log statements different", you seem to have a different intent here, though. An intent I didn't suspect, especially in the light of the group discussions where everybody except yourself voiced concerns about the verbosity of the fluent API variant you wanted to switch to...

Maybe this was our main misunderstanding here.

Anyway, this PR is approved now, so feel free to merge it.
